{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a03daf8b-0582-4700-8186-ef2660d70f45",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Analysis\n",
    "\n",
    "This notebook includes code for analyzing and plotting processed library activity measurements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "397fa83e-0457-46d5-9b66-4853f1079107",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "os.environ[\"POLARS_MAX_THREADS\"] = os.environ[\"SLURM_CPUS_PER_TASK\"]\n",
    "print(f\"{os.environ['SLURM_CPUS_PER_TASK']} thread(s) available\")\n",
    "\n",
    "FIGURE_EXPORT_TYPE = \"svg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "934a1c09-73c2-4054-8ba3-24ed1c0c1e48",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from IPython.display import display, Markdown, HTML\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib_inline.backend_inline\n",
    "matplotlib_inline.backend_inline.set_matplotlib_formats(\"jpeg\")\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as pyplot\n",
    "import seaborn\n",
    "from matplotlib_venn import venn2\n",
    "from statannotations.Annotator import Annotator\n",
    "\n",
    "import numpy\n",
    "import polars\n",
    "import pandas\n",
    "import scipy\n",
    "import statsmodels.api as statsmods\n",
    "\n",
    "import json\n",
    "\n",
    "polars.toggle_string_cache(True)\n",
    "_ = polars.Config.set_tbl_rows(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c179cf5-df56-432c-a83b-4adb442a22f3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Style Matplotlib plots\n",
    "\"\"\"\n",
    "\n",
    "plot_style = {\n",
    "    \"font.size\": 12,\n",
    "    \"font.family\": \"sans-serif\",\n",
    "    \"font.sans-serif\": [\"Inter\"],\n",
    "    \"figure.figsize\": [2, 2],\n",
    "    \"figure.dpi\": 150,\n",
    "    \"savefig.dpi\": 300,\n",
    "    \"figure.facecolor\": (1, 1, 1, 0),\n",
    "    \"text.usetex\": False,\n",
    "    \"lines.markersize\": 3,\n",
    "    \"axes.titleweight\": \"bold\",\n",
    "    \"axes.labelweight\": 600,\n",
    "    \"axes.labelsize\": 9,\n",
    "    \"axes.facecolor\": \"none\",\n",
    "    \"axes.spines.top\": False,\n",
    "    \"axes.spines.right\": False,\n",
    "    \"axes.linewidth\": 1.5,\n",
    "    \"grid.color\": \"#303030\",\n",
    "    \"grid.alpha\": 0.1,\n",
    "    \"xtick.labelsize\": 9,\n",
    "    \"xtick.major.width\": 1.5,\n",
    "    \"xtick.minor.width\": 1,\n",
    "    \"xtick.minor.visible\": False,\n",
    "    \"ytick.labelsize\": 9,\n",
    "    \"ytick.major.width\": 1.5,\n",
    "    \"ytick.minor.width\": 1,\n",
    "    \"ytick.minor.visible\": False,\n",
    "    \"figure.frameon\": False,\n",
    "    \"legend.frameon\": False,\n",
    "    \"legend.fancybox\": False,\n",
    "    \"legend.fontsize\": 9,\n",
    "    \"legend.scatterpoints\": 1,\n",
    "    \"legend.markerscale\": 1,\n",
    "    \"legend.handlelength\": 1.0,\n",
    "    \"legend.handletextpad\": 0.2,\n",
    "    \"axes.prop_cycle\": matplotlib.cycler(color=[\"k\", \"b\", \"g\", \"r\", \"c\", \"y\"]),\n",
    "    \"hatch.linewidth\": 4,\n",
    "}\n",
    "\n",
    "pyplot.style.use(plot_style)\n",
    "_ = pyplot.ioff()\n",
    "\n",
    "cm_white_to_gray = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_gray\", [\"#ffffff\", \"#404040\"])\n",
    "cm_white_to_red = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_red\", [\"#ffffff\", \"#ff0000\"])\n",
    "cm_white_to_blue = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_blue\", [\"#ffffff\", \"#0000ff\"])\n",
    "cm_white_to_green = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_green\", [\"#ffffff\", \"#009000\"])\n",
    "\n",
    "cm_light_to_gray = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_gray\", [\"#e0e0e0\", \"#404040\"])\n",
    "cm_light_to_red = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_red\", [\"#ffe0e0\", \"#ff0000\"])\n",
    "cm_light_to_blue = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_blue\", [\"#e0e0ff\", \"#0000ff\"])\n",
    "cm_light_to_green = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_green\", [\"#e0ffe0\", \"#009000\"])\n",
    "\n",
    "boxplot_style_args = {\n",
    "    \"whiskerprops\": {\n",
    "        \"color\": \"#000000\"\n",
    "    },\n",
    "    \"boxprops\": {\n",
    "        \"edgecolor\": \"#000000\",\n",
    "    },\n",
    "    \"flierprops\": {\n",
    "        \"marker\": \"o\",\n",
    "        \"markerfacecolor\": \"#000000\",\n",
    "        \"markersize\": 1,\n",
    "        \"markeredgecolor\": None\n",
    "    },\n",
    "    \"medianprops\": {\n",
    "        \"color\": \"#000000\"\n",
    "    },\n",
    "    \"capprops\": {\n",
    "        \"color\": \"#000000\"\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "124061fb-dca8-426d-9a18-e7ee0d2170e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Common plotting functions\n",
    "\"\"\"\n",
    "\n",
    "# Adapted from https://stackoverflow.com/a/53865762\n",
    "\n",
    "def calc_weights_for_scatter(x, y, bins=20):\n",
    "    if type(x) is not numpy.ndarray or type(y) is not numpy.ndarray:\n",
    "        raise ValueError(\"x and y must be `numpy.ndarray`s\")\n",
    "    \n",
    "    bin_weights, xbins, ybins = numpy.histogram2d(x, y, bins=bins, density=False)\n",
    "    weights = scipy.interpolate.interpn(points=(0.5*(xbins[1:] + xbins[:-1]), 0.5*(ybins[1:] + ybins[:-1])), values=bin_weights, xi=numpy.vstack([x, y]).T, method=\"splinef2d\", bounds_error=False)\n",
    "    weights[numpy.where(numpy.isnan(weights))] = 0.0\n",
    "    \n",
    "    normalizer = matplotlib.colors.Normalize(vmin=numpy.min(weights), vmax=numpy.max(weights))\n",
    "    \n",
    "    return weights, normalizer\n",
    "\n",
    "def reorder_points_and_weights(x, y, w, f = None):\n",
    "    if type(x) is not numpy.ndarray or type(y) is not numpy.ndarray or type(w) is not numpy.ndarray:\n",
    "        raise ValueError(\"x, y, and w must be `numpy.ndarray`s\")\n",
    "    \n",
    "    x_, y_, w_ = x, y, w\n",
    "    if f is not None:\n",
    "        x_, y_, w_ = f(x), f(y), f(w)\n",
    "    order = numpy.argsort(w_)\n",
    "    return {\n",
    "        \"x\": x_[order],\n",
    "        \"y\": y_[order],\n",
    "        \"c\": w_[order]\n",
    "    }\n",
    "\n",
    "NAME_TRANSLATION_TABLE = {\n",
    "    \"WT\": \"+/+\",\n",
    "    \"Rhet\": \"R90W/+\",\n",
    "    \"Rhom\": \"R90W/R90W\",\n",
    "    \"Ehet\": \"E168d2/+\",\n",
    "    \"Ehom\": \"E168d2/E168d2\",\n",
    "    \"CrxKO\": \"-/-\",\n",
    "    \"chip_peak_id\": \"CRE\"\n",
    "}\n",
    "\n",
    "def translate_genotype_to_label(genotype):\n",
    "    return NAME_TRANSLATION_TABLE[genotype]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "159feded-edd2-45b0-94e5-f3153b144ac1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Load library metadata\n",
    "\"\"\"\n",
    "\n",
    "library_metadata = polars.read_csv(\"Library_Details/library_metadata.tsv\", sep=\"\\t\", dtypes={\n",
    "    \"variant_type\": polars.Categorical\n",
    "})\n",
    "\n",
    "chip_peak_metadata = polars.read_csv(\"Library_Details/chip_peak_metadata.tsv\", sep=\"\\t\", dtypes={\n",
    "    \"atac_crx_dependent\": polars.Categorical,\n",
    "    \"epigenome_group\": polars.Categorical\n",
    "})\n",
    "\n",
    "display(Markdown(\"### Library Metadata\"))\n",
    "display(library_metadata)\n",
    "\n",
    "display(Markdown(\"### ChIP Peak Metadata\"))\n",
    "display(chip_peak_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d975eb-1844-41bc-9be0-3d69b350ff63",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Read in processed counts, and join with library metadata\n",
    "\"\"\"\n",
    "\n",
    "processed_samples_counts = polars.read_parquet(\"MPRA/processed_library_counts.parquet\")\n",
    "\n",
    "processed_samples_counts = processed_samples_counts.join(library_metadata, on=\"library_id\", how=\"left\").join(chip_peak_metadata, on=\"chip_peak_id\", how=\"left\")\n",
    "\n",
    "display(processed_samples_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d789a905-cab9-4578-a8ee-587e57d950ca",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# GLOBAL PATTERNS OF REPORTER EXPRESSION ACROSS GENOTYPES\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beccd01e-e936-434a-91e4-b1f473082071",
   "metadata": {},
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Scatterplot correlation of CRE activity\"))\n",
    "    \n",
    "    for variant_type, cmap, base_color in [(\"wildtype\", cm_light_to_blue, \"#0000ff\"), (\"allCrxSites\", cm_light_to_red, \"#ff0000\")]:\n",
    "    \n",
    "        data = library_df.filter(polars.col(\"variant_type\") == variant_type).with_columns([\n",
    "            polars.col(\"activity_mean\") + 2e-5\n",
    "        ]).sort([\"genotype\", \"library_id\"]).pivot(values=\"activity_mean\", index=\"library_id\", columns=\"genotype\", maintain_order=True).to_pandas()\n",
    "\n",
    "        scrambled_data = library_df.filter(polars.col(\"variant_type\") == \"scrambled\").with_columns([\n",
    "            polars.col(\"activity_mean\") + 2e-5\n",
    "        ]).sort([\"genotype\", \"library_id\"]).pivot(values=\"activity_mean\", index=\"library_id\", columns=\"genotype\", maintain_order=True).to_pandas()\n",
    "        \n",
    "        #display(data)\n",
    "        #display(scrambled_data)\n",
    "\n",
    "        genotype_comparisons = [\n",
    "            (\"WT\", \"Rhet\"),\n",
    "            (\"WT\", \"Ehet\"),\n",
    "            (\"WT\", \"Rhom\"),\n",
    "            (\"WT\", \"Ehom\"),\n",
    "            (\"WT\", \"CrxKO\"),\n",
    "            (\"Rhet\", \"Rhom\"),\n",
    "            (\"Ehet\", \"Ehom\"),\n",
    "            (\"Rhet\", \"Ehet\"),\n",
    "            (\"Rhom\", \"Ehom\")\n",
    "        ]\n",
    "\n",
    "        figure, axes = pyplot.subplots(nrows=2, ncols=5, sharex=False, sharey=False, figsize=(15, 6), layout=\"compressed\")\n",
    "        bounds = (1e-5, 1e3)\n",
    "        \n",
    "        for index, ((genotype1, genotype2), axis) in enumerate(zip(genotype_comparisons, axes.flat)):\n",
    "\n",
    "            # Define grid of bins for evaulating heatmap and lines\n",
    "            bins = numpy.geomspace(start=bounds[0], stop=bounds[1], num=50)\n",
    "\n",
    "            x = data[genotype1].to_numpy()\n",
    "            y = data[genotype2].to_numpy()\n",
    "            \n",
    "            # Compute robust regression of scrambled sequences\n",
    "            scrambled_x = scrambled_data[genotype1].to_numpy()\n",
    "            scrambled_y = scrambled_data[genotype2].to_numpy()\n",
    "            log_scrambled_x_with_intercept = statsmods.add_constant(numpy.log10(scrambled_x))\n",
    "\n",
    "            rlm_results = statsmods.RLM(numpy.log10(scrambled_y), log_scrambled_x_with_intercept, M=statsmods.robust.norms.TrimmedMean(), missing=\"raise\").fit()\n",
    "            intercept, slope = rlm_results.params\n",
    "            axis.plot(bins, numpy.power(10, rlm_results.predict(statsmods.add_constant(numpy.log10(bins)), transform=False)), color=\"#00000070\", linestyle=(0, (5, 10)), linewidth=0.5, marker=None, label=f\"m={slope:.2f}\")\n",
    "            \n",
    "            data_rlm_residuals = numpy.abs(numpy.log10(y) - rlm_results.predict(statsmods.add_constant(numpy.log10(x))))\n",
    "            data_rlm_residuals_bound = numpy.percentile(data_rlm_residuals, 80)\n",
    "            axis.plot(bins, numpy.power(10, rlm_results.predict(statsmods.add_constant(numpy.log10(bins)), transform=False) + data_rlm_residuals_bound), color=base_color + \"70\", linestyle=(0, (2, 2)), linewidth=0.5, marker=None, label=\"80th %ile residual\" if index == 0 else \"_nolegend_\")\n",
    "            axis.plot(bins, numpy.power(10, rlm_results.predict(statsmods.add_constant(numpy.log10(bins)), transform=False) - data_rlm_residuals_bound), color=base_color + \"70\", linestyle=(0, (2, 2)), linewidth=0.5, marker=None, label=\"_nolegend_\")\n",
    "\n",
    "            # Calculate R^2\n",
    "            fit = scipy.stats.linregress(numpy.log10(x), numpy.log10(y))\n",
    "            fit_func = lambda x: fit.slope * x + fit.intercept\n",
    "            \n",
    "            # Plot wildtype CREs as a scatterplot colored by density\n",
    "            weights, color_normalizer = calc_weights_for_scatter(x, y, bins=bins)\n",
    "            axis.scatter(**reorder_points_and_weights(x, y, weights), cmap=cmap, norm=color_normalizer, s=6, marker=\"o\", rasterized=True, label=f\"R²={fit.rvalue**2:.2f}\")\n",
    "\n",
    "            # Plot scrambled sequences\n",
    "            axis.scatter(scrambled_data[genotype1], scrambled_data[genotype2], marker=\"+\", color=\"#000000\", s=6, alpha=1, linewidth=0.5, label=\"scrambled\" if index == 0 else \"_nolegend_\", rasterized=True)\n",
    "            \n",
    "            # Plot basal\n",
    "            g1_basal_mean = library_df.filter((polars.col(\"genotype\") == genotype1) & (polars.col(\"variant_type\") == \"basal\"))[\"activity_mean\"][0]\n",
    "            g2_basal_mean = library_df.filter((polars.col(\"genotype\") == genotype2) & (polars.col(\"variant_type\") == \"basal\"))[\"activity_mean\"][0]\n",
    "            axis.scatter(g1_basal_mean, g2_basal_mean, marker=\"+\", color=\"#ff7000\", s=16, linewidth=1, label=\"basal\" if index == 0 else \"_nolegend_\")\n",
    "            \n",
    "            # Plot out-of-bounds\n",
    "            axis.add_patch(matplotlib.patches.Polygon(\n",
    "                xy=[(bounds[1], 2e-5), (2e-5, 2e-5), (2e-5, bounds[1]), (0, bounds[1]), (0, 0), (bounds[1], 0)],\n",
    "                facecolor=\"#00000000\",\n",
    "                edgecolor=\"#00000008\",\n",
    "                hatch=r\"//\"\n",
    "            ))\n",
    "            \n",
    "            # Format plot\n",
    "            axis.set(xlim=bounds, ylim=bounds, xlabel=translate_genotype_to_label(genotype1), ylabel=translate_genotype_to_label(genotype2))\n",
    "            axis.set_xscale(\"log\", base=10)\n",
    "            axis.set_yscale(\"log\", base=10)\n",
    "            axis.xaxis.set_major_locator(pyplot.LogLocator(numticks=5))\n",
    "            axis.yaxis.set_major_locator(pyplot.LogLocator(numticks=5))\n",
    "            axis.tick_params(which=\"major\", bottom=True, top=False, left=True, right=False, labelbottom=True, labeltop=False, labelleft=True, labelright=False)\n",
    "            axis.tick_params(which=\"minor\", bottom=False, top=False, left=False, right=False)\n",
    "            \n",
    "            axis.set_aspect(aspect=\"equal\", adjustable=None, anchor=\"C\")\n",
    "\n",
    "            handles, labels = axis.get_legend_handles_labels()\n",
    "            legend = axis.legend(handles, labels, loc=\"upper left\")\n",
    "\n",
    "        for axis in axes.flat[len(genotype_comparisons):]:\n",
    "            axis.axis(\"off\")\n",
    "            \n",
    "        figure.savefig(f\"Figures/expression_across_genotypes_{variant_type}_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "        display(Markdown(f\"### {variant_type}\"))\n",
    "        display(figure)\n",
    "        pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f1ff73f-7b43-44ee-8df0-64c1c939a091",
   "metadata": {},
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Heatmap correlation of CRE classification\"))\n",
    "    \n",
    "    for variant_type, cmap in [(\"wildtype\", cm_white_to_blue), (\"allCrxSites\", cm_white_to_red)]:\n",
    "        \n",
    "        genotype_comparisons = [\n",
    "            (\"WT\", \"Rhet\"),\n",
    "            (\"WT\", \"Ehet\"),\n",
    "            (\"WT\", \"Rhom\"),\n",
    "            (\"WT\", \"Ehom\"),\n",
    "            (\"WT\", \"CrxKO\"),\n",
    "            (\"Rhet\", \"Rhom\"),\n",
    "            (\"Ehet\", \"Ehom\"),\n",
    "            (\"Rhet\", \"Ehet\"),\n",
    "            (\"Rhom\", \"Ehom\")\n",
    "        ]\n",
    "\n",
    "        figure, axes = pyplot.subplots(nrows=2, ncols=5, sharex=False, sharey=False, figsize=(20, 8), layout=\"compressed\")\n",
    "\n",
    "        activity_classes = [\"strong_silencer\", \"weak_silencer\", \"inactive\", \"weak_enhancer\", \"strong_enhancer\"]\n",
    "\n",
    "        for (genotype1, genotype2), axis in zip(genotype_comparisons, axes.flat):\n",
    "            raw_intersections = {}\n",
    "            for genotype1_activity_class in activity_classes:\n",
    "                raw_intersections[genotype1_activity_class] = {}\n",
    "                for genotype2_activity_class in reversed(activity_classes):\n",
    "                    genotype1_activity_class_ids = set(\n",
    "                        library_df.filter(\n",
    "                            (polars.col(\"genotype\") == genotype1) &\n",
    "                            (polars.col(\"variant_type\") == variant_type) &\n",
    "                            (polars.col(\"activity_class\") == genotype1_activity_class)\n",
    "                        )[\"chip_peak_id\"]\n",
    "                    )\n",
    "                    genotype2_activity_class_ids = set(\n",
    "                        library_df.filter(\n",
    "                            (polars.col(\"genotype\") == genotype2) &\n",
    "                            (polars.col(\"variant_type\") == variant_type) &\n",
    "                            (polars.col(\"activity_class\") == genotype2_activity_class)\n",
    "                        )[\"chip_peak_id\"]\n",
    "                    )\n",
    "                    intersection = genotype1_activity_class_ids & genotype2_activity_class_ids\n",
    "                    raw_intersections[genotype1_activity_class][genotype2_activity_class] = len(intersection)\n",
    "            intersections = pandas.DataFrame(raw_intersections)\n",
    "            seaborn.heatmap(intersections, ax=axis, cmap=cmap, square=True, annot=True, fmt=\"d\", cbar=False)\n",
    "            axis.set(xlabel=translate_genotype_to_label(genotype1), ylabel=translate_genotype_to_label(genotype2))\n",
    "            \n",
    "            axis.set_aspect(\"equal\", adjustable=None, anchor=\"C\")\n",
    "\n",
    "        for axis in axes.flat[len(genotype_comparisons):]:\n",
    "            axis.axis(\"off\")\n",
    "\n",
    "        figure.savefig(f\"Figures/activity_class_intersections_across_genotypes_{variant_type}_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "        display(Markdown(f\"### {variant_type}\"))\n",
    "        display(figure)\n",
    "        pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "482dba88-be6d-47aa-801b-8b2765404664",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Scatterplot correlation of wt CRE vs mutant CRE activity\"))\n",
    "\n",
    "    genotypes = [\n",
    "        \"WT\",\n",
    "        \"Rhet\",\n",
    "        \"Ehet\",\n",
    "        \"Rhom\",\n",
    "        \"Ehom\",\n",
    "        \"CrxKO\"\n",
    "    ]\n",
    "\n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=6, sharex=False, sharey=True, figsize=(18, 4), layout=\"compressed\")\n",
    "    x_bounds = (1e-5, 1e3)\n",
    "    y_bounds = (1e-7, 1e4)\n",
    "    \n",
    "    first_axis = axes.flat[0]\n",
    "\n",
    "    for index, (genotype, axis) in enumerate(zip(genotypes, axes.flat)):\n",
    "        \n",
    "        data = library_df.filter((polars.col(\"genotype\") == genotype) & ((polars.col(\"variant_type\") == \"wildtype\") | (polars.col(\"variant_type\") == \"allCrxSites\"))).with_columns([\n",
    "            polars.col(\"activity_mean\") + 2e-5\n",
    "        ]).pivot(values=\"activity_mean\", index=\"chip_peak_id\", columns=\"variant_type\").with_columns([\n",
    "            (polars.col(\"allCrxSites\")/polars.col(\"wildtype\")).alias(\"foldChange\")\n",
    "        ]).to_pandas()\n",
    "        \n",
    "        #display(data)\n",
    "\n",
    "        # Define grid of bins for evaulating heatmap and lines\n",
    "        x_bins = numpy.geomspace(start=x_bounds[0], stop=y_bounds[1], num=50)\n",
    "        y_bins = numpy.geomspace(start=y_bounds[0], stop=y_bounds[1], num=50)\n",
    "\n",
    "        x = data[\"wildtype\"].to_numpy()\n",
    "        y = data[\"foldChange\"].to_numpy()\n",
    "\n",
    "        # Calculate R^2\n",
    "        fit = scipy.stats.linregress(numpy.log10(x), numpy.log10(y))\n",
    "        fit_func = lambda x: fit.slope * x + fit.intercept\n",
    "\n",
    "        # Plot wildtype CREs as a scatterplot colored by density\n",
    "        weights, color_normalizer = calc_weights_for_scatter(x, y, bins=(x_bins, y_bins))\n",
    "        axis.scatter(**reorder_points_and_weights(x, y, weights), cmap=cm_light_to_gray, norm=color_normalizer, s=6, marker=\"o\", rasterized=True, label=f\"R²={fit.rvalue**2:.2f}\")\n",
    "\n",
    "        # Plot basal\n",
    "        basal_mean = library_df.filter((polars.col(\"genotype\") == genotype) & (polars.col(\"variant_type\") == \"basal\"))[\"activity_mean\"][0]\n",
    "        axis.scatter(basal_mean, 1, marker=\"+\", color=\"#ff7000\", s=16, linewidth=1, label=\"basal\" if index == 0 else \"_nolegend_\")\n",
    "        axis.axvline(x=basal_mean, color=\"#ff700080\", linewidth=1.5, linestyle=\":\")\n",
    "        axis.axhline(y=1, color=\"#ff700080\", linewidth=1.5, linestyle=\":\")\n",
    "        \n",
    "        # Plot out-of-bounds        \n",
    "        axis.add_patch(matplotlib.patches.Polygon(\n",
    "            xy=[(1e3, 2e-5/1e3), (2e-5, 1), (2e-5, 1e4), (0, 1e4), (0, 0)],\n",
    "            facecolor=\"#00000000\",\n",
    "            edgecolor=\"#00000008\",\n",
    "            hatch=r\"//\"\n",
    "        ))\n",
    "        \n",
    "        # Format plot\n",
    "        axis.set(xlim=x_bounds, ylim=y_bounds, title=translate_genotype_to_label(genotype), xlabel=\"Intact motif activity\", ylabel=\"Killed motif fold change\")\n",
    "        axis.xaxis.label.set_color(\"#0000ff\")\n",
    "        axis.yaxis.label.set_color(\"#ff0000\")\n",
    "        axis.set_xscale(\"log\", base=10)\n",
    "        axis.set_yscale(\"log\", base=10)\n",
    "        axis.xaxis.set_major_locator(pyplot.LogLocator(numticks=5))\n",
    "        axis.yaxis.set_major_locator(pyplot.LogLocator(numticks=7))\n",
    "        axis.tick_params(which=\"major\", bottom=True, top=False, left=True, right=False, labelbottom=True, labeltop=False, labelleft=True, labelright=False)\n",
    "        axis.tick_params(which=\"minor\", bottom=False, top=False, left=False, right=False)\n",
    "        axis.set_aspect(aspect=\"equal\", adjustable=None, anchor=\"C\")\n",
    "\n",
    "        handles, labels = axis.get_legend_handles_labels()\n",
    "        legend = axis.legend(handles, labels, loc=\"upper left\")\n",
    "\n",
    "    figure.savefig(f\"Figures/expression_wt_vs_crxMut_motifs_{variant_type}_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(Markdown(f\"### {variant_type}\"))\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788f42fc-b88a-4fbe-ad5c-ee5e8f1f3a6b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Heatmap correlation of wt CRE vs mutant CRE classification\"))\n",
    "    \n",
    "    genotypes = [\n",
    "        \"WT\",\n",
    "        \"Rhet\",\n",
    "        \"Ehet\",\n",
    "        \"Rhom\",\n",
    "        \"Ehom\",\n",
    "        \"CrxKO\"\n",
    "    ]\n",
    "    \n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=6, sharex=False, sharey=False, figsize=(24, 4), layout=\"compressed\")\n",
    "    \n",
    "    for index, (genotype, axis) in enumerate(zip(genotypes, axes.flat)):\n",
    "        activity_classes = [\"strong_silencer\", \"weak_silencer\", \"inactive\", \"weak_enhancer\", \"strong_enhancer\"]\n",
    "\n",
    "        raw_intersections = {}\n",
    "        for wtMotif_activity_class in activity_classes:\n",
    "            raw_intersections[wtMotif_activity_class] = {}\n",
    "            for mutMotif_activity_class in reversed(activity_classes):\n",
    "                wtMotif_activity_class_ids = set(\n",
    "                    library_df.filter(\n",
    "                        (polars.col(\"genotype\") == genotype) &\n",
    "                        (polars.col(\"variant_type\") == \"wildtype\") &\n",
    "                        (polars.col(\"activity_class\") == wtMotif_activity_class)\n",
    "                    )[\"chip_peak_id\"]\n",
    "                )\n",
    "                mutMotif_activity_class_ids = set(\n",
    "                    library_df.filter(\n",
    "                        (polars.col(\"genotype\") == genotype) &\n",
    "                        (polars.col(\"variant_type\") == \"allCrxSites\") &\n",
    "                        (polars.col(\"activity_class\") == mutMotif_activity_class)\n",
    "                    )[\"chip_peak_id\"]\n",
    "                )\n",
    "                \n",
    "                intersection = wtMotif_activity_class_ids & mutMotif_activity_class_ids\n",
    "                raw_intersections[wtMotif_activity_class][mutMotif_activity_class] = len(intersection)\n",
    "        intersections = pandas.DataFrame(raw_intersections)\n",
    "        seaborn.heatmap(intersections, ax=axis, cmap=cm_white_to_gray, square=True, annot=True, fmt=\"d\", cbar=False)\n",
    "        axis.set(xlabel=\"Intact Motif CREs\", ylabel=\"Killed Motif CREs\", title=f\"{genotype}\")\n",
    "        axis.xaxis.label.set_color(\"#0000ff\")\n",
    "        axis.yaxis.label.set_color(\"#ff0000\")\n",
    "\n",
    "        axis.set_aspect(\"equal\", adjustable=None, anchor=\"C\")\n",
    "\n",
    "    for axis in axes.flat[len(genotypes):]:\n",
    "        axis.axis(\"off\")\n",
    "\n",
    "    figure.savefig(f\"Figures/activity_class_intersections_wt_vs_crxMut_motifs_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(Markdown(f\"### {variant_type}\"))\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f49a500-82b9-4725-9e94-bbf515dcf1a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Heatmap of CRE activity across all genotypes\"))\n",
    "    \n",
    "    processed_data = library_df.filter(polars.col(\"variant_type\") == \"wildtype\")\n",
    "\n",
    "    heatmap_data = processed_data.with_columns([\n",
    "        polars.col(\"activity_mean\") + 2e-5\n",
    "    ]).pivot(values=\"activity_mean\", index=\"chip_peak_id\", columns=\"genotype\").select([\n",
    "        \"chip_peak_id\",\n",
    "        \"WT\",\n",
    "        \"Rhet\",\n",
    "        \"Ehet\",\n",
    "        \"Rhom\",\n",
    "        \"Ehom\",\n",
    "        \"CrxKO\"\n",
    "    ]).sort(polars.col(\"CrxKO\")/polars.col(\"WT\"), descending=True).rename(NAME_TRANSLATION_TABLE)\n",
    "    \n",
    "    activity_df = heatmap_data.to_pandas().set_index(\"CRE\")\n",
    "    display(activity_df)\n",
    "    \n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=1, figsize=(1.5, 6))\n",
    "    \n",
    "    axes = seaborn.heatmap(axes=axes, data=activity_df, cmap=\"PiYG\", norm=matplotlib.colors.LogNorm(vmin=1e-4, vmax=1e3, clip=True), yticklabels=False, rasterized=True, cbar_kws={\"fraction\": 0.08})\n",
    "    axes.collections[0].colorbar.ax.set_ylabel(\"Transcriptional Activity\", rotation=270, labelpad=10, fontsize=10)\n",
    "    figure.savefig(f\"Figures/element_activity_heatmap_across_genotypes_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c96fbcae-9b11-491a-b29b-55289e539b35",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "\n",
    "# DEREPRESSED SILENCERS\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "639cd369-bda0-490f-9fbf-2d3925cf37e2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "element_subgroups = {}\n",
    "\n",
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    \n",
    "    if library_name not in element_subgroups:\n",
    "        element_subgroups[library_name] = {}\n",
    "    \n",
    "    element_subgroups[library_name][\"wt_silencers\"] = library_df.filter(\n",
    "        (polars.col(\"variant_type\") == \"wildtype\") &\n",
    "        (polars.col(\"genotype\") == \"WT\") &\n",
    "        ((polars.col(\"activity_class\") == \"strong_silencer\") | (polars.col(\"activity_class\") == \"weak_silencer\"))\n",
    "    )[\"chip_peak_id\"]\n",
    "    \n",
    "    element_subgroups[library_name][\"wt_strong_enhancers\"] = library_df.filter(\n",
    "        (polars.col(\"variant_type\") == \"wildtype\") &\n",
    "        (polars.col(\"genotype\") == \"WT\") &\n",
    "        (polars.col(\"activity_class\") == \"strong_enhancer\")\n",
    "    )[\"chip_peak_id\"]\n",
    "    \n",
    "    wt_vs_e168d2 = library_df.filter(\n",
    "        (polars.col(\"variant_type\") == \"wildtype\") &\n",
    "        (polars.col(\"genotype\") == \"WT\")\n",
    "    ).select([\n",
    "        polars.col(\"chip_peak_id\"),\n",
    "        polars.col(\"activity_class\").alias(\"activity_class_WT\")\n",
    "    ]).join(\n",
    "        library_df.filter(\n",
    "            (polars.col(\"variant_type\") == \"wildtype\") &\n",
    "            (polars.col(\"genotype\") == \"Ehom\")\n",
    "        ).select([\n",
    "            polars.col(\"chip_peak_id\"),\n",
    "            polars.col(\"activity_class\").alias(\"activity_class_Ehom\")\n",
    "        ]), how=\"left\", on=\"chip_peak_id\"\n",
    "    )\n",
    "    \n",
    "    element_subgroups[library_name][\"e168d2_derepressed_silencers\"] = wt_vs_e168d2.filter(\n",
    "        ((polars.col(\"activity_class_WT\") == \"strong_silencer\") | (polars.col(\"activity_class_WT\") == \"weak_silencer\")) &\n",
    "        ((polars.col(\"activity_class_Ehom\") == \"weak_enhancer\") | (polars.col(\"activity_class_Ehom\") == \"strong_enhancer\"))\n",
    "    )[\"chip_peak_id\"]\n",
    "    \n",
    "    element_subgroups[library_name][\"e168d2_always_silencers\"] = wt_vs_e168d2.filter(\n",
    "        ((polars.col(\"activity_class_WT\") == \"strong_silencer\") | (polars.col(\"activity_class_WT\") == \"weak_silencer\")) &\n",
    "        ((polars.col(\"activity_class_Ehom\") == \"strong_silencer\") | (polars.col(\"activity_class_Ehom\") == \"weak_silencer\"))\n",
    "    )[\"chip_peak_id\"]\n",
    "    \n",
    "    element_subgroups[library_name][\"e168d2_always_enhancers\"] = wt_vs_e168d2.filter(\n",
    "        ((polars.col(\"activity_class_WT\") == \"weak_enhancer\") | (polars.col(\"activity_class_WT\") == \"strong_enhancer\")) &\n",
    "        ((polars.col(\"activity_class_Ehom\") == \"weak_enhancer\") | (polars.col(\"activity_class_Ehom\") == \"strong_enhancer\"))\n",
    "    )[\"chip_peak_id\"]\n",
    "    \n",
    "    element_subgroups[library_name][\"e168d2_always_strong_enhancers\"] = wt_vs_e168d2.filter(\n",
    "        (polars.col(\"activity_class_WT\") == \"strong_enhancer\") &\n",
    "        (polars.col(\"activity_class_Ehom\") == \"strong_enhancer\")\n",
    "    )[\"chip_peak_id\"]\n",
    "    \n",
    "    element_subgroups[library_name][\"crxko_derepressed_silencers\"] = library_df.filter(\n",
    "        (polars.col(\"variant_type\") == \"wildtype\") &\n",
    "        (polars.col(\"genotype\") == \"WT\")\n",
    "    ).select([\n",
    "        \"chip_peak_id\",\n",
    "        polars.col(\"activity_class\").alias(\"activity_class_WT\")\n",
    "    ]).join(\n",
    "        library_df.filter(\n",
    "            (polars.col(\"variant_type\") == \"wildtype\") &\n",
    "            (polars.col(\"genotype\") == \"CrxKO\")\n",
    "        ).select([\n",
    "            \"chip_peak_id\",\n",
    "            polars.col(\"activity_class\").alias(\"activity_class_CrxKO\")\n",
    "        ]), how=\"left\", on=\"chip_peak_id\"\n",
    "    ).filter(\n",
    "        ((polars.col(\"activity_class_WT\") == \"strong_silencer\") | (polars.col(\"activity_class_WT\") == \"weak_silencer\")) &\n",
    "        ((polars.col(\"activity_class_CrxKO\") == \"weak_enhancer\") | (polars.col(\"activity_class_CrxKO\") == \"strong_enhancer\"))\n",
    "    )[\"chip_peak_id\"]\n",
    "\n",
    "for library_name in [\"rho\", \"hsp68\"]:\n",
    "    \n",
    "    display(Markdown(f\"## [{library_name}] Subgroup intersections\"))\n",
    "    \n",
    "    subgroup_comparisons = [\n",
    "        (\"e168d2_derepressed_silencers\", \"crxko_derepressed_silencers\"),\n",
    "        (\"wt_strong_enhancers\", \"e168d2_always_enhancers\"),\n",
    "        (\"wt_silencers\", \"e168d2_always_silencers\"),\n",
    "    ]\n",
    "    \n",
    "    for (subgroup1, subgroup2) in subgroup_comparisons:\n",
    "        display(Markdown(f\"### {subgroup1} and {subgroup2}\"))\n",
    "\n",
    "        figure, axes = pyplot.subplots(nrows=1, ncols=1, figsize=(4, 3))\n",
    "\n",
    "        venn = venn2([set(element_subgroups[library_name][subgroup1]), set(element_subgroups[library_name][subgroup2])], ax=axes)\n",
    "        venn.get_label_by_id(\"A\").set_text(subgroup1)\n",
    "        venn.get_label_by_id(\"B\").set_text(subgroup2)\n",
    "\n",
    "        venn.patches[0].set_facecolor(\"#29C732\")\n",
    "        venn.patches[0].set_alpha(1.0)\n",
    "\n",
    "        venn.patches[1].set_facecolor(\"#FF9AB6\")\n",
    "        venn.patches[1].set_alpha(1.0)\n",
    "\n",
    "        venn.patches[2].set_facecolor(\"#94B174\")\n",
    "        venn.patches[2].set_alpha(1.0)\n",
    "\n",
    "        figure.savefig(f\"Figures/overlap_{subgroup1}_vs_{subgroup2}_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "        display(figure)\n",
    "        pyplot.close(figure)\n",
    " \n",
    "for library in element_subgroups:\n",
    "    display(Markdown(f\"## [{library}] Subgroup sizes\"))\n",
    "    list_string = \"\"\n",
    "    for subgroup in element_subgroups[library]:\n",
    "        matching_CRE_sequences = library_metadata.filter(\n",
    "            (polars.col(\"chip_peak_id\").is_in(element_subgroups[library][subgroup])) &\n",
    "            (polars.col(\"variant_type\") == \"wildtype\")\n",
    "        ).select([\"chip_peak_id\", \"CRE_sequence\"])\n",
    "        output_file_name = f\"Motif_Analysis/{subgroup}_{library}.fa\"\n",
    "        with open(output_file_name, \"w\") as out:\n",
    "            for (chip_peak_id, CRE_sequence) in matching_CRE_sequences.iter_rows():\n",
    "                out.write(f\">{chip_peak_id}\\n{CRE_sequence}\\n\")\n",
    "        \n",
    "        list_string += f\" - {subgroup}: {len(element_subgroups[library][subgroup])} elements (wrote CRE sequences to `{output_file_name}`)\\n\"\n",
    "        \n",
    "    display(Markdown(list_string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7a03e14-0499-4181-912a-6d31a6365cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Activity of subgroups\"))\n",
    "    \n",
    "    basal_data = library_df.filter(\n",
    "        (polars.col(\"library_id\") == \"basal\")\n",
    "    ).with_columns([\n",
    "        polars.col(\"activity_mean\") + 2e-5\n",
    "    ]).with_columns([\n",
    "        polars.col(\"genotype\").apply(translate_genotype_to_label)\n",
    "    ]).select([\n",
    "        \"genotype\",\n",
    "        \"activity_mean\",\n",
    "    ]).to_pandas()\n",
    "    \n",
    "    for subgroup_name in [\"e168d2_derepressed_silencers\", \"e168d2_always_silencers\", \"e168d2_always_enhancers\"]:\n",
    "    \n",
    "        display(Markdown(f\"### {subgroup_name}\"))\n",
    "\n",
    "        processed_data = library_df.filter(\n",
    "            ((polars.col(\"variant_type\") == \"wildtype\") | (polars.col(\"variant_type\") == \"allCrxSites\")) &\n",
    "            polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][subgroup_name])\n",
    "        ).with_columns([\n",
    "            polars.col(\"activity_mean\") + 2e-5\n",
    "        ]).with_columns([\n",
    "            polars.col(\"genotype\").apply(translate_genotype_to_label)\n",
    "        ]).to_pandas()\n",
    "\n",
    "        figure, axes = pyplot.subplots(nrows=1, ncols=1, figsize=(2.5, 3))\n",
    "\n",
    "        plot_data_params = dict(\n",
    "            data=processed_data,\n",
    "            x=\"genotype\",\n",
    "            y=\"activity_mean\",\n",
    "            hue=\"variant_type\",\n",
    "            order=[\"+/+\", \"R90W/+\", \"E168d2/+\", \"R90W/R90W\", \"E168d2/E168d2\", \"-/-\"],\n",
    "            hue_order=[\"wildtype\", \"allCrxSites\"],\n",
    "        )\n",
    "        _ = seaborn.boxplot(ax=axes, **plot_data_params, notch=False, palette=[\"#6060ff\", \"#ff3030\"], linewidth=1, **boxplot_style_args)\n",
    "\n",
    "        _ = seaborn.swarmplot(ax=axes, data=basal_data, x=\"genotype\", y=\"activity_mean\", order=[\"+/+\", \"R90W/+\", \"E168d2/+\", \"R90W/R90W\", \"E168d2/E168d2\", \"-/-\"], s=5, color=\"#ff7000\")\n",
    "\n",
    "        # Plot out-of-bounds\n",
    "        axes.add_patch(matplotlib.patches.Polygon(\n",
    "            xy=[(-0.5, 2e-5), (len(plot_data_params[\"order\"]) - 0.5, 2e-5), (len(plot_data_params[\"order\"]) - 0.5, 0), (0, 0)],\n",
    "            facecolor=\"#00000000\",\n",
    "            edgecolor=\"#00000008\",\n",
    "            hatch=r\"//\"\n",
    "        ))\n",
    "\n",
    "        axes.set(ylim=(1e-5, 1e2), xlabel=None, ylabel=\"Transcriptional Activity\")\n",
    "        axes.set_yscale(\"log\")\n",
    "        axes.tick_params(axis=\"x\", rotation=90)\n",
    "\n",
    "        figure.savefig(f\"Figures/subgroup_activity_{subgroup_name}_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "        display(figure)\n",
    "        pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fc279d4-c453-4e74-9ecb-d22f62106575",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# COOPERATIVITY OF CRX SITES\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83fb4a6a-5c43-41e4-9fdd-d25e98b8f119",
   "metadata": {},
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.sort([\"library\"]).groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Cooperativity of CREs\"))\n",
    "    \n",
    "    data = library_df.with_columns([\n",
    "        polars.col(\"activity_mean\") + 2e-5\n",
    "    ]).filter(polars.col(\"num_crx_sites\") == 2).sort([\"chip_peak_id\", \"genotype\", \"variant_type\"]).with_columns([\n",
    "        polars.when(polars.col(\"library_id\").str.ends_with(\"_1_double\")).then(1).when(polars.col(\"library_id\").str.ends_with(\"2_double\")).then(2).otherwise(None).alias(\"single_site_num\"),\n",
    "    ])\n",
    "    \n",
    "    processed_data = data.filter(polars.col(\"variant_type\") == \"wildtype\").join(\n",
    "        data.filter((polars.col(\"variant_type\") == \"singleCrxSite\") & (polars.col(\"single_site_num\") == 1)),\n",
    "        how=\"inner\", on=[\"genotype\", \"chip_peak_id\"], suffix=\"_single1\"\n",
    "    ).join(\n",
    "        data.filter((polars.col(\"variant_type\") == \"singleCrxSite\") & (polars.col(\"single_site_num\") == 2)),\n",
    "        how=\"inner\", on=[\"genotype\", \"chip_peak_id\"], suffix=\"_single2\"\n",
    "    ).join(\n",
    "        data.filter((polars.col(\"variant_type\") == \"allCrxSites\")),\n",
    "        how=\"inner\", on=[\"genotype\", \"chip_peak_id\"], suffix=\"_allSites\"\n",
    "    ).with_columns([\n",
    "        #((polars.col(\"activity_mean_allSites\") - polars.col(\"activity_mean\")) - (polars.col(\"activity_mean_single1\") - polars.col(\"activity_mean\")) - (polars.col(\"activity_mean_single2\") - polars.col(\"activity_mean\"))).alias(\"double_single_activity_difference\")\n",
    "        #((polars.col(\"activity_mu_allSites\") - polars.col(\"activity_mu\")) - (polars.col(\"activity_mu_single1\") - polars.col(\"activity_mu\")) - (polars.col(\"activity_mu_single2\") - polars.col(\"activity_mu\"))).fill_nan(0).alias(\"double_single_activity_difference\")\n",
    "        (((polars.col(\"activity_mean\") - polars.col(\"activity_mean_single1\")) + (polars.col(\"activity_mean\") - polars.col(\"activity_mean_single2\")))/(polars.col(\"activity_mean\") - polars.col(\"activity_mean_allSites\"))).fill_nan(0).alias(\"double_single_activity_difference\")\n",
    "    ])\n",
    "    \n",
    "    heatmap_data = processed_data.pivot(values=\"double_single_activity_difference\", index=\"chip_peak_id\", columns=\"genotype\").select([\n",
    "        \"chip_peak_id\",\n",
    "        \"WT\",\n",
    "        \"Rhet\",\n",
    "        \"Ehet\",\n",
    "        \"Rhom\",\n",
    "        \"Ehom\",\n",
    "        \"CrxKO\"\n",
    "    ]).sort(polars.col(\"CrxKO\")/polars.col(\"WT\"), descending=True).rename(NAME_TRANSLATION_TABLE).to_pandas().set_index(\"CRE\")\n",
    "    display(heatmap_data)\n",
    "    \n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=1, figsize=(1.5, 3))\n",
    "    \n",
    "    axes = seaborn.heatmap(axes=axes, data=heatmap_data, cmap=\"PiYG\", norm=matplotlib.colors.LogNorm(vmin=1e-3, vmax=1e3), yticklabels=False, rasterized=True, cbar_kws={\"fraction\": 0.08, \"label\": \"Transcriptional Activity\"})\n",
    "    axes.collections[0].colorbar.ax.set_ylabel(\"(Δa$_{site1}$ + Δa$_{site2}$) / (Δa$_{site1+2}$)\", rotation=270, labelpad=10, fontsize=10)\n",
    "    figure.savefig(f\"Figures/cooperativity_across_genotypes_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82435229-9240-478e-8359-a73d9ec0bf05",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# CHROMATIN STATE\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9505ce1e-381e-4ffb-8e85-d032c5a0faae",
   "metadata": {},
   "outputs": [],
   "source": [
    "for library_name, library_df in processed_samples_counts.filter((polars.col(\"variant_type\") == \"wildtype\") & (polars.col(\"genotype\") == \"WT\")).sort(\"library\").groupby(\"library\", maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] Fraction of Activity Classifications by chromatin annotation of CRE genomic origin in +/+\"))\n",
    "    \n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=1, figsize=(5.5, 3.5))\n",
    "    \n",
    "    data = library_df.with_columns([\n",
    "        polars.col(\"activity_mean\") + 2e-5,\n",
    "        polars.when(polars.col(\"activity_class\") == \"strong_enhancer\").then(4)\n",
    "        .when(polars.col(\"activity_class\") == \"weak_enhancer\").then(3)\n",
    "        .when(polars.col(\"activity_class\") == \"inactive\").then(2)\n",
    "        .when(polars.col(\"activity_class\") == \"weak_silencer\").then(1)\n",
    "        .when(polars.col(\"activity_class\") == \"strong_silencer\").then(0).alias(\"activity_class_rank\")\n",
    "    ]).sort([\"epigenome_group\", \"atac_crx_dependent\", \"activity_class_rank\"]).groupby([\"epigenome_group\", \"atac_crx_dependent\", \"activity_class\"], maintain_order=True).agg([\n",
    "        polars.count().alias(\"num_CREs\")\n",
    "    ]).with_columns([\n",
    "        polars.sum(\"num_CREs\").over([\"epigenome_group\", \"atac_crx_dependent\"]).alias(\"total_CREs\")\n",
    "    ]).with_columns([\n",
    "        (polars.col(\"num_CREs\")/polars.col(\"total_CREs\")).alias(\"fraction_CREs\")\n",
    "    ])\n",
    "    \n",
    "    x_axis_labels = []\n",
    "    \n",
    "    prior_row = None\n",
    "    for (activity_class_name, group_data), color in zip(data.groupby(\"activity_class\", maintain_order=True), seaborn.diverging_palette(280, 120, l=80, n=5, center=\"light\")):\n",
    "        if prior_row is None:\n",
    "            axes.barh(y=group_data[\"epigenome_group\"]+\"_\"+group_data[\"atac_crx_dependent\"], width=group_data[\"fraction_CREs\"], label=f\"{activity_class_name}\", facecolor=color)\n",
    "            prior_row = group_data[\"fraction_CREs\"]\n",
    "        else:\n",
    "            axes.barh(y=group_data[\"epigenome_group\"]+\"_\"+group_data[\"atac_crx_dependent\"], width=group_data[\"fraction_CREs\"], label=f\"{activity_class_name}\", left=prior_row, facecolor=color)\n",
    "            prior_row += group_data[\"fraction_CREs\"]\n",
    "    \n",
    "    axes.set(xlim=(0, 1), xlabel=\"Fraction\", ylabel=\"Chromatin Group\")\n",
    "    axes.legend(loc=\"right\", title=f\"Activity Class\", facecolor=\"#ffffff\", bbox_to_anchor=(1.4,0.5))\n",
    "    axes.invert_yaxis()\n",
    "    \n",
    "    figure.savefig(f\"Figures/fraction_activity_classes_by_epigenome_group_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e84379cb-2893-43a3-8ffb-74b25fae3e04",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for library_name in [\"rho\", \"hsp68\"]:\n",
    "    display(Markdown(f\"## [{library_name}] Derepressed silencer epigenome group classifications\"))\n",
    "    \n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=1, figsize=(5.5, 0.6))\n",
    "    \n",
    "    data = chip_peak_metadata.filter(\n",
    "        polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][\"e168d2_derepressed_silencers\"])\n",
    "    ).sort([\"epigenome_group\"]).groupby([\"epigenome_group\"], maintain_order=True).agg([\n",
    "        polars.count().alias(\"num_CREs\")\n",
    "    ]).with_columns([\n",
    "        (polars.col(\"num_CREs\")/polars.sum(\"num_CREs\")).alias(\"fraction_CREs\")\n",
    "    ])\n",
    "    \n",
    "    x_axis_labels = []\n",
    "    \n",
    "    prior_row = None\n",
    "    for (activity_class_name, group_data), color in zip(data.groupby(\"epigenome_group\", maintain_order=True), seaborn.husl_palette(3)):\n",
    "        if prior_row is None:\n",
    "            axes.barh(y=[\"Derepressed Silencers\"], width=group_data[\"fraction_CREs\"], label=f\"{activity_class_name}\", facecolor=color)\n",
    "            prior_row = group_data[\"fraction_CREs\"]\n",
    "        else:\n",
    "            axes.barh(y=[\"Derepressed Silencers\"], width=group_data[\"fraction_CREs\"], label=f\"{activity_class_name}\", left=prior_row, facecolor=color)\n",
    "            prior_row += group_data[\"fraction_CREs\"]\n",
    "    \n",
    "    axes.set(xlim=(0, 1), ylabel=None, xlabel=\"Fraction\")\n",
    "    axes.legend(loc=\"right\", title=f\"Activity Class\", facecolor=\"#ffffff\", bbox_to_anchor=(1.6, 0.5))\n",
    "    \n",
    "    figure.savefig(f\"Figures/fraction_activity_classes_derepressed_silencers_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e932e1d0-39eb-4c6f-9e95-b928f34f95e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (library_name, epigenome_group_name), library_df in processed_samples_counts.filter((polars.col(\"variant_type\") == \"wildtype\")).sort([\"library\", \"epigenome_group\"]).groupby([\"library\", \"epigenome_group\"], maintain_order=True):\n",
    "    display(Markdown(f\"## [{library_name}] CRX-dependent accessibility of CRE genomic origin (Group {epigenome_group_name})\"))\n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=1, figsize=(5.5, 4))\n",
    "    \n",
    "    data = library_df.with_columns([\n",
    "        polars.col(\"activity_mean\") + 2e-5,\n",
    "        polars.col(\"genotype\").apply(translate_genotype_to_label),\n",
    "    ]).to_pandas()\n",
    "    \n",
    "    basal_data = processed_samples_counts.filter(\n",
    "        (polars.col(\"library_id\") == \"basal\")\n",
    "        & (polars.col(\"library\") == library_name)\n",
    "    ).with_columns([\n",
    "        polars.col(\"activity_mean\") + 2e-5\n",
    "    ]).with_columns([\n",
    "        polars.col(\"genotype\").apply(translate_genotype_to_label)\n",
    "    ]).select([\n",
    "        \"genotype\",\n",
    "        \"activity_mean\",\n",
    "        polars.lit(\"dependent\").alias(\"atac_crx_dependent\")\n",
    "    ]).to_pandas()\n",
    "    \n",
    "    plot_data_params = dict(\n",
    "        data=data,\n",
    "        x=\"atac_crx_dependent\",\n",
    "        y=\"activity_mean\",\n",
    "        hue=\"genotype\",\n",
    "        order=[\"dependent\", \"independent\"],\n",
    "        hue_order=[\"+/+\", \"R90W/+\", \"E168d2/+\", \"R90W/R90W\", \"E168d2/E168d2\", \"-/-\"]\n",
    "    )\n",
    "    _ = seaborn.boxplot(ax=axes, **plot_data_params, notch=False, palette=seaborn.husl_palette(6), linewidth=1, dodge=True,  **boxplot_style_args)\n",
    "    \n",
    "    _ = seaborn.stripplot(ax=axes, data=basal_data, x=plot_data_params[\"x\"], y=plot_data_params[\"y\"], hue=plot_data_params[\"hue\"], order=plot_data_params[\"order\"], hue_order=plot_data_params[\"hue_order\"], palette=seaborn.husl_palette(6), s=5, color=\"#ff7000\")\n",
    "    \n",
    "    # Plot out-of-bounds\n",
    "    axes.add_patch(matplotlib.patches.Polygon(\n",
    "        xy=[(-0.5, 2e-5), (len(plot_data_params[\"order\"]) - 0.5, 2e-5), (len(plot_data_params[\"order\"]) - 0.5, 0), (0, 0)],\n",
    "        facecolor=\"#00000000\",\n",
    "        edgecolor=\"#00000008\",\n",
    "        hatch=r\"//\"\n",
    "    ))\n",
    "    \n",
    "    axes.set(ylim=(1e-5, 3e2), xlabel=\"ATAC Accessibility\", ylabel=\"Transcriptional Activity\")\n",
    "    axes.set_yscale(\"log\")\n",
    "    axes.legend(loc=\"right\", title=f\"Group {epigenome_group_name}\", facecolor=\"#ffffff\", bbox_to_anchor=(1.6,0.5))\n",
    "    \n",
    "    figure.savefig(f\"Figures/activity_by_atac_crx_dependence_group{epigenome_group_name}_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbf62b90-e75a-4671-b189-59c357afc2c1",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# INFORMATION CONTENT ANALYSIS\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa625ee-a8f9-4c3c-892b-b8f3b88ef0b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "information_content = polars.read_csv(\"Library_Details/library_information_content.tsv\", sep=\"\\t\")\n",
    "display(information_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede2b452-4fc4-4127-9a53-001c1fba4643",
   "metadata": {},
   "outputs": [],
   "source": [
    "for library_name in [\"rho\", \"hsp68\"]:\n",
    "    display(Markdown(f\"## [{library_name}] Subgroup information content and occupancy\"))\n",
    "    \n",
    "    data = library_metadata.filter(polars.col(\"variant_type\") == \"wildtype\").with_columns([\n",
    "        polars.when(polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][\"e168d2_derepressed_silencers\"])).then(\"E168d2 Derepressed Silencers\")\n",
    "        .when(polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][\"wt_strong_enhancers\"])).then(\"+/+ Strong Enhancers\")\n",
    "        .when(polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][\"wt_silencers\"]) & polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][\"e168d2_derepressed_silencers\"]).is_not()).then(\"+/+ Silencers\")\n",
    "        .otherwise(None).alias(\"derepression_status\")\n",
    "    ]).drop_nulls(\"derepression_status\").join(information_content, on=\"library_id\", how=\"left\").sort(\"derepression_status\")\n",
    "    \n",
    "    data = data.to_pandas()\n",
    "    data[\"variant_type\"].cat = data[\"variant_type\"].cat.remove_unused_categories()\n",
    "    \n",
    "    subplot_details = [\n",
    "        (\"info_content\", \"Information Content\"),\n",
    "        (\"predicted_occupancy_CRX\", \"Predicted CRX Occupancy\"),\n",
    "        (\"total_occupancy\", \"Total Occupancy\")\n",
    "    ]\n",
    "    \n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=len(subplot_details), figsize=(2.35*len(subplot_details), 3.5))\n",
    "    \n",
    "    for (yvar, ylabel), axis in zip(subplot_details, axes.flat):\n",
    "        plot_data_params = dict(\n",
    "            data=data,\n",
    "            x=\"derepression_status\",\n",
    "            y=yvar,\n",
    "            order=[\"+/+ Strong Enhancers\", \"E168d2 Derepressed Silencers\", \"+/+ Silencers\"],\n",
    "        )\n",
    "        _ = seaborn.boxplot(ax=axis, **plot_data_params, notch=False, palette=[\"#6060ff\"], linewidth=1, dodge=True, **boxplot_style_args)\n",
    "\n",
    "        axis.set(xlabel=None, ylabel=ylabel)\n",
    "        axis.tick_params(axis=\"x\", rotation=20)\n",
    "        for label in axis.get_xticklabels():\n",
    "            label.set_ha(\"right\")\n",
    "        axis.legend([],[], frameon=False)\n",
    "    \n",
    "        display(Markdown(f\"### {ylabel}\"))\n",
    "        annotator = Annotator(axis, **plot_data_params, pairs=(\n",
    "            (\"+/+ Strong Enhancers\", \"+/+ Silencers\"),\n",
    "            (\"+/+ Strong Enhancers\", \"E168d2 Derepressed Silencers\"),\n",
    "            (\"+/+ Silencers\", \"E168d2 Derepressed Silencers\"),\n",
    "        ))\n",
    "        annotator.configure(test=\"Mann-Whitney\", verbose=True)\n",
    "        annotator.apply_test()\n",
    "        annotator.annotate()\n",
    "\n",
    "    figure.tight_layout()\n",
    "    figure.savefig(f\"Figures/information_content_and_occupancy_E168d2_derepressed_silencers_{library_name}.{FIGURE_EXPORT_TYPE}\", bbox_inches=\"tight\")\n",
    "    display(figure)\n",
    "    pyplot.close(figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df546ce2-54d3-494e-9819-eb71cf8d21cc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import statsmodels\n",
    "\n",
    "bcor_overlap = polars.read_csv(\"BCOR/CRX_BCOR_overlap.tsv\", sep=\"\\t\").with_columns([\n",
    "    ((polars.col(\"nearby_BCOR_in_P0\") > 0)).alias(\"overlaps_BCOR\")\n",
    "])\n",
    "\n",
    "display(Markdown(\"## CRX ChIP Peak Overlap with BCOR Cut&Run\"))\n",
    "display(bcor_overlap)\n",
    "\n",
    "# column value is the number of replicates (0, 1, or 2) in which a BCOR peak overlapped the given CRX ChIP peak\n",
    "\n",
    "comparisons = [\n",
    "    (\"e168d2_derepressed_silencers\", \"wt_silencers\"),\n",
    "]\n",
    "\n",
    "for library_name in [\"rho\", \"hsp68\"]:\n",
    "    display(Markdown(f\"### {library_name}\"))\n",
    "    \n",
    "    for subgroup1, subgroup2 in comparisons:\n",
    "        display(Markdown(f\"#### {subgroup1} vs. {subgroup2}\"))\n",
    "        \n",
    "        num_elems_1 = len(element_subgroups[library_name][subgroup1])\n",
    "        num_bcor_overlapping_1 = bcor_overlap.filter(polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][subgroup1]) & polars.col(\"overlaps_BCOR\")).height\n",
    "            \n",
    "        num_elems_2 = len(element_subgroups[library_name][subgroup2])\n",
    "        num_bcor_overlapping_2 = bcor_overlap.filter(polars.col(\"chip_peak_id\").is_in(element_subgroups[library_name][subgroup2]) & polars.col(\"overlaps_BCOR\")).height\n",
    "\n",
    "        print(f\"{num_bcor_overlapping_1}/{num_elems_1} ({num_bcor_overlapping_1/num_elems_1:0.2%}) vs. {num_bcor_overlapping_2}/{num_elems_2} ({num_bcor_overlapping_2/num_elems_2:0.2%})\")\n",
    "        \n",
    "        zstat, pval = statsmodels.stats.proportion.proportions_ztest([num_bcor_overlapping_1, num_bcor_overlapping_2], [num_elems_1, num_elems_2])\n",
    "        \n",
    "        print(f\"z={zstat:.2e}, p={pval:.2e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56eab9ae-ae9d-4496-8416-c963ed13a9ad",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
