{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dafb82aa-8e01-4381-a723-45c7997fc71f",
   "metadata": {},
   "source": [
    "# Information Content Analysis\n",
    "\n",
    "Processes library metadata to calculate information content and occupancy for each library member, using code derived from Friedman et al. (eLife 2021)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a0c074-98f3-452a-80d8-554e9eb71626",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display, Markdown, HTML\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "import numpy\n",
    "import pandas\n",
    "import scipy\n",
    "\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6849925-5a05-4567-890f-8b92589a4666",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Supporting Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2ac5a18-14fe-4076-a93a-0a5cfe33c11f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Courtesy of Ryan Friedman\n",
    "\n",
    "def rev_comp(seq):\n",
    "    \"\"\"Take the reverse compliment of a sequence\n",
    "    Parameters\n",
    "    ----------\n",
    "    seq : str\n",
    "        The original sequence.\n",
    "    Returns\n",
    "    -------\n",
    "    new_seq : str\n",
    "        The reverse compliment.\n",
    "    \"\"\"\n",
    "    compliment = {\"A\": \"T\", \"C\": \"G\", \"G\": \"C\", \"T\": \"A\"}\n",
    "    new_seq = seq[::-1]\n",
    "    new_seq = \"\".join([compliment[i] for i in new_seq])\n",
    "    return new_seq\n",
    "\n",
    "def peek(fin):\n",
    "    \"\"\" Peek at the next line in a file.\n",
    "    Parameters\n",
    "    ----------\n",
    "    fin : file input stream\n",
    "    Returns\n",
    "    -------\n",
    "    line : str\n",
    "    \"\"\"\n",
    "    pos = fin.tell()\n",
    "    line = fin.readline()\n",
    "    fin.seek(pos)\n",
    "    return line\n",
    "\n",
    "\n",
    "def gobble(fin):\n",
    "    \"\"\"Gobble up lines in the file until we have reached the start of a motif or EOF.\n",
    "    Parameters\n",
    "    ----------\n",
    "    fin : file input stream\n",
    "    Returns\n",
    "    -------\n",
    "    lines : str\n",
    "        The lines that got gobbled, including newline characters.\n",
    "    \"\"\"\n",
    "    lines = \"\"\n",
    "    while True:\n",
    "        line = peek(fin)\n",
    "        if len(line) == 0 or line[:5] == \"MOTIF\":\n",
    "            break\n",
    "        else:\n",
    "            lines += fin.readline()\n",
    "\n",
    "    return lines\n",
    "\n",
    "\n",
    "def read_pwm_files(filename):\n",
    "    \"\"\"Given a MEME file, read in all PWMs. PWMs are stored as DataFrames, and the list of PWMs is represented as a\n",
    "    Series, where keys are primary motif identifiers and values are the DataFrames.\n",
    "    Parameters\n",
    "    ----------\n",
    "    filename : str\n",
    "        Name of the file to read in.\n",
    "    Returns\n",
    "    -------\n",
    "    pwm_ser : pandas.Series\n",
    "        The list of PWMs parsed from the file.\n",
    "    \"\"\"\n",
    "    pwm_ser = {}\n",
    "    with open(filename) as fin:\n",
    "        # Lines before the first motif is encountered\n",
    "        gobble(fin)\n",
    "\n",
    "        # Do-while like behavior to read in the data\n",
    "        # Do <read in motif> while not EOF\n",
    "        while True:\n",
    "            # MOTIF <motif name> [alternate ID]\n",
    "            motif_id = fin.readline().split()[1]\n",
    "\n",
    "            # Empty line\n",
    "            fin.readline()\n",
    "            # \"letter-probability matrix: [other info]\"\n",
    "            fin.readline()\n",
    "\n",
    "            # Every line that starts with a space is a new position in the PWM, if the first character is not a space\n",
    "            # it is not part of the PWM.\n",
    "            pwm = []\n",
    "            while peek(fin)[0] == \" \":\n",
    "                pwm.append(fin.readline().split())\n",
    "\n",
    "            # Make a DataFrame and add to the list\n",
    "            pwm = pandas.DataFrame(pwm, dtype=float, columns=[\"A\", \"C\", \"G\", \"T\"])\n",
    "            pwm_ser[motif_id] = pwm\n",
    "            # Read up any extra info such as the URL\n",
    "            gobble(fin)\n",
    "\n",
    "            # Check if EOF\n",
    "            if len(peek(fin)) == 0:\n",
    "                break\n",
    "\n",
    "    pwm_ser = pandas.Series(pwm_ser)\n",
    "    return pwm_ser\n",
    "\n",
    "\n",
    "def ewm_from_letter_prob(pwm_df, pseudocount=0.0001, rt=2.5):\n",
    "    \"\"\"Compute an energy weight matrix from a letter probability matrix. Normalize the PWM to the maximum letter\n",
    "    probability at each position and then compute relative free energies using the formula ddG = -RT ln(p_b,i / p_c,\n",
    "    i), where p_b,i is the probability of base b, p_c,i is the probability of the consensus base, and ddG is relative\n",
    "    free energy.\n",
    "    Parameters\n",
    "    ----------\n",
    "    pwm_df : pandas.DataFrame\n",
    "        The letter probability matrix, where each row is a position of the motif and columns represent A, C, G, T.\n",
    "    pseudocount : float\n",
    "        Pseudocount value to add to every value to account for zeros in the PWM.\n",
    "    rt : float\n",
    "        The value of RT to use in the formula in kJ/mol.\n",
    "    Returns\n",
    "    -------\n",
    "    ewm_df : pandas.DataFrame\n",
    "        The weight matrix of free energies relatives to the consensus sequence.\n",
    "    \"\"\"\n",
    "    pwm_df = pwm_df.copy()\n",
    "    pwm_df += pseudocount\n",
    "    # Normalize each position by the most frequent letter to get relative Kd\n",
    "    pwm_df = pwm_df.apply(lambda x: x / x.max(), axis=1)\n",
    "    # Convert to EWM\n",
    "    ewm_df = -rt * numpy.log(pwm_df)\n",
    "    ewm_df.columns = [\"A\", \"C\", \"G\", \"T\"]\n",
    "    return ewm_df\n",
    "\n",
    "\n",
    "def ewm_to_dict(ewm):\n",
    "    \"\"\"Convert a DataFrame representation of an EWM to a dictionary for faster indexing.\n",
    "    Parameters\n",
    "    ----------\n",
    "    ewm : pandas.DataFrame\n",
    "    Returns\n",
    "    -------\n",
    "    ewm_dict : {int: {str: float}}\n",
    "        Dictionary of dictionaries, where the outer keys are positions, the inner keys are letters, and the values\n",
    "        are values of the matrix\n",
    "    \"\"\"\n",
    "    ewm_dict = ewm.to_dict(orient=\"index\")\n",
    "    return ewm_dict\n",
    "\n",
    "\n",
    "def read_pwm_to_ewm(filename, pseudocount=0.0001, rt=2.5):\n",
    "    \"\"\"Read in a file of letter probability matrices, convert them to EWMs, and then convert the DataFrames to\n",
    "    dictionaries for faster indexting.\n",
    "    Parameters\n",
    "    ----------\n",
    "    filename : str\n",
    "        Name of the file to read in.\n",
    "     pseudocount : float\n",
    "        Pseudocount value to add to every value to account for zeros in the PWM.\n",
    "    rt : float\n",
    "        The value of RT to use in the formula in kJ/mol.\n",
    "    Returns\n",
    "    -------\n",
    "    ewm_dict : {int: {str: float}}\n",
    "        Dictionary of dictionaries, where the outer keys are positions, the inner keys are letters, and the values\n",
    "        are values of the matrix\n",
    "    \"\"\"\n",
    "    # Wrapper function handle to convert each PWM to an EWM\n",
    "    # Read in the file\n",
    "    pwms = read_pwm_files(filename)\n",
    "    # Convert to EWM dicts\n",
    "    ewms = pwms.apply(ewm_from_letter_prob, args=(pseudocount, rt)).apply(ewm_to_dict)\n",
    "    return ewms\n",
    "\n",
    "\n",
    "def energy_landscape(seq, ewm):\n",
    "    \"\"\"Scans both strands of a sequence with energy matrix\n",
    "    Parameters\n",
    "    ----------\n",
    "    seq : str\n",
    "        The sequence to scan.\n",
    "    ewm : dict {int: {str: float}}\n",
    "        Dictionary of dictionaries, where the outer keys are positions, the inner keys are letters, and the values\n",
    "        are delta delta G relative to the consensus sequence.\n",
    "    Returns\n",
    "    -------\n",
    "    fscores, rscores: numpy.array, dtype=float\n",
    "        Represents the EWM scores for each subsequence on the forward and reverse strand.\n",
    "    \"\"\"\n",
    "    motif_len = len(ewm.keys())\n",
    "    # Number of positions where the motif can be scored\n",
    "    n_scores = len(seq) - motif_len + 1\n",
    "    fscores = numpy.zeros(n_scores)\n",
    "    # Reverse compliment scores\n",
    "    rscores = fscores.copy()\n",
    "    r_seq = rev_comp(seq)\n",
    "\n",
    "    # Calculate occ for forward and reverse k-mer at every position\n",
    "    for pos in range(n_scores):\n",
    "        f_kmer = seq[pos:pos + motif_len]\n",
    "        r_kmer = r_seq[pos:pos + motif_len]\n",
    "\n",
    "        # Initialize energy score\n",
    "        fscore = 0\n",
    "        rscore = 0\n",
    "        \n",
    "        # This is faster than using the enumerate function\n",
    "        # Calculate the EWM score for the k-mer starting at pos\n",
    "        for i in range(motif_len):\n",
    "            fscore += ewm[i][f_kmer[i]]\n",
    "            rscore += ewm[i][r_kmer[i]]\n",
    "\n",
    "        fscores[pos] = fscore\n",
    "        rscores[pos] = rscore\n",
    "\n",
    "    # rscores needs to be reversed so the indexing corresponds to the appropriate position in the original sequence (\n",
    "    # i.e. just the compliment, not the reverse compliment)\n",
    "    rscores = rscores[::-1]\n",
    "\n",
    "    return fscores, rscores\n",
    "\n",
    "\n",
    "def occupancy_landscape(seq, ewm, mu):\n",
    "    \"\"\"Compute the occupancy landscape by scanning sequence with the energy matrix and then calculate the relative\n",
    "    free energy for each k-mer subsequence on the forward and reverse strand at chemical potential mu.\n",
    "    Parameters\n",
    "    ----------\n",
    "    seq : str\n",
    "        The sequence to scan.\n",
    "    ewm : dict {int: {str: float}}\n",
    "        Dictionary of dictionaries, where the outer keys are positions, the inner keys are letters, and the values\n",
    "        are delta delta G relative to the consensus sequence.\n",
    "    mu : int\n",
    "        Chemical potential of the TF\n",
    "    Returns\n",
    "    -------\n",
    "    fscores, rscores: numpy.array, dtype=float\n",
    "        Represents the occupancy scores for each subsequence on the forward and reverse strand.\n",
    "    \"\"\"\n",
    "    fscores, rscores = energy_landscape(seq, ewm)\n",
    "    # Convert EWM scores to occupancies\n",
    "    fscores = 1 / (1 + numpy.exp(fscores - mu))\n",
    "    rscores = 1 / (1 + numpy.exp(rscores - mu))\n",
    "    return fscores, rscores\n",
    "\n",
    "\n",
    "def total_landscape(seq, ewms, mu):\n",
    "    \"\"\"Compute the occupancy landscape for each TF and join it all together into a DataFrame. Pad the ends of the\n",
    "    positional information so every TF occupancy landscape is the same length.\n",
    "    Parameters\n",
    "    ----------\n",
    "    seq : str\n",
    "        The DNA sequence.\n",
    "    ewms : pandas.Series or dict {str: {int: {str: float}}}\n",
    "        Keys/index are TF names and values are dictionary representations of the EWMs.\n",
    "    mu : int or float\n",
    "        TF chemical potential.\n",
    "    Returns\n",
    "    -------\n",
    "    landscape : pandas.DataFrame, dtype=float\n",
    "        The occupancy of each TF at each position in each orientation. Rows are positions, columns are TFs and\n",
    "        orientations, values indicate the predicted occupancy starting at the position.\n",
    "    \"\"\"\n",
    "    landscape = {}\n",
    "    seq_len = len(seq)\n",
    "    # For each TF\n",
    "    for name, ewm in ewms.items():\n",
    "        # Get the predicted occupancy and add it to the list\n",
    "        fscores, rscores = occupancy_landscape(seq, ewm, mu)\n",
    "        landscape[f\"{name}_F\"] = fscores\n",
    "        landscape[f\"{name}_R\"] = rscores\n",
    "\n",
    "    # Pad the ends of the lists to the length of the sequence\n",
    "    for key, val in landscape.items():\n",
    "        amount_to_add = seq_len - len(val)\n",
    "        landscape[key] = numpy.pad(val, (0, amount_to_add), mode=\"constant\", constant_values=0)\n",
    "\n",
    "    landscape = pandas.DataFrame(landscape)\n",
    "    return landscape\n",
    "\n",
    "\n",
    "def total_occupancy(seq, ewms, mu):\n",
    "    \"\"\"For each TF, calculate its predicted occupancy over the sequence given the energy matrix and chemical\n",
    "    potential. Then, summarize the information as the total occupancy of each TF over the entire sequence.\n",
    "    Parameters\n",
    "    ----------\n",
    "    seq : str\n",
    "        The DNA sequence.\n",
    "    ewms : pandas.Series or dict {str: {int: {str: float}}}\n",
    "        Keys/index are TF names and values are dictionary representations of the EWMs.\n",
    "    mu : int or float\n",
    "        TF chemical potential.\n",
    "    Returns\n",
    "    -------\n",
    "    occ_profile : pandas.Series, dtype=float\n",
    "        The total occupancy profile of each TF on the sequence.\n",
    "    \"\"\"\n",
    "    occ_landscape = total_landscape(seq, ewms, mu)\n",
    "    occ_profile = {}\n",
    "    # Add together F and R strand\n",
    "    if type(ewms) is dict:\n",
    "        keys = ewms.keys()\n",
    "    else:\n",
    "        keys = ewms.index\n",
    "    for tf in keys:\n",
    "        occ_profile[tf] = occ_landscape[[f\"{tf}_F\", f\"{tf}_R\"]].sum().sum()\n",
    "\n",
    "    occ_profile = pandas.Series(occ_profile)\n",
    "    return occ_profile\n",
    "\n",
    "def information_content(occupancies, diversity_cutoff=0.5, log=numpy.log2):\n",
    "    \"\"\"Given a list of TF occupancies, compute total occupancy, diversity, and information content.\n",
    "    Parameters\n",
    "    ----------\n",
    "    occupancies : pandas.Series\n",
    "        Predicted occupancy for a collection of TFs on a given sequence.\n",
    "    diversity_cutoff : float\n",
    "        Cutoff to call a TF \"occupied\" on the sequence.\n",
    "    log : Function handle\n",
    "        Function to use for computing the log. Default is log2 so information content is in bits, natural log should be used for\n",
    "        biophysical applications.\n",
    "    Returns\n",
    "    -------\n",
    "    result : pandas.Series\n",
    "        The total occupancy, diversity, and information content of the provided sequence.\n",
    "    \"\"\"\n",
    "    # Calculate total occupancy of all TFs on the sequence\n",
    "    total_occ = occupancies.sum()\n",
    "    # Count how many of the TFs are occupied, i.e. have motifs present in the sequence\n",
    "    diversity = (occupancies > diversity_cutoff).sum()\n",
    "    # Since the occupancies are continuous values, we need to use the Gamma function to compute entropy. Gamma(n+1)=n!\n",
    "    # W = N! / prod(N_i!)\n",
    "    microstates = scipy.special.gamma(total_occ + 1) / (occupancies + 1).apply(scipy.special.gamma).product()\n",
    "    # S = log W\n",
    "    info_content = log(microstates)\n",
    "\n",
    "    result = pandas.Series({\n",
    "        \"total_occupancy\": total_occ,\n",
    "        \"diversity\": diversity,\n",
    "        \"info_content\": info_content\n",
    "    })\n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef2e7886-8c1d-4fd0-9b37-44aa793fab18",
   "metadata": {},
   "source": [
    "## Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0687ce53-0724-4a0a-a8d0-285d42f9232b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Load TF EWMs\n",
    "\"\"\"\n",
    "\n",
    "ewm_ser = read_pwm_to_ewm(\"photoreceptorAndEnrichedMotifs.meme\")\n",
    "display(ewm_ser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f73b36f3-8f03-4e96-b6d9-7260603bc87b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Load library metadata, and nearby genes TSS\n",
    "\"\"\"\n",
    "\n",
    "library_metadata = pandas.read_csv(\"library_metadata.tsv\", sep=\"\\t\")[[\"library_id\", \"CRE_sequence\"]]\n",
    "display(library_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e9c269-5b56-47b1-bd62-89fa14a05d59",
   "metadata": {},
   "outputs": [],
   "source": [
    "mu = 9\n",
    "\n",
    "raw_information_contents = {}\n",
    "for sequence in tqdm(library_metadata[\"CRE_sequence\"].str.upper()):\n",
    "    occ_ser = total_occupancy(sequence, ewm_ser, mu)\n",
    "    for motif, pred_occ in occ_ser.to_dict().items():\n",
    "        short_motif = motif.split(\"_\")[0]\n",
    "        key = f\"predicted_occupancy_{short_motif}\"\n",
    "        if key not in raw_information_contents:\n",
    "            raw_information_contents[key] = []\n",
    "        raw_information_contents[key].append(pred_occ)\n",
    "    inf_cont = information_content(occ_ser).to_dict()\n",
    "    for key in inf_cont:\n",
    "        if key not in raw_information_contents:\n",
    "            raw_information_contents[key] = []\n",
    "        raw_information_contents[key].append(inf_cont[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e368c18d-e070-4123-b3d0-c14996217dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in raw_information_contents:\n",
    "    library_metadata[key] = raw_information_contents[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de7dacf8-8689-449e-aaa4-522ca36deab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "807132a8-d14b-498a-ab08-a30bbaea8f64",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_metadata.drop([\"CRE_sequence\"], axis=1).to_csv(\"library_information_content.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f4c3c6-602a-493a-983c-b072b1e45e4d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
