{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "46dbead1-ea77-40ba-99d4-dd26662748a5",
   "metadata": {},
   "source": [
    "# Count Processing\n",
    "\n",
    "This notebook includes code for processing counts generated by `count_bcs.sh`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bb81557-ddca-45df-947f-42329d1e85a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "os.environ[\"POLARS_MAX_THREADS\"] = os.environ[\"SLURM_CPUS_PER_TASK\"]\n",
    "print(f\"{os.environ['SLURM_CPUS_PER_TASK']} thread(s) available\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f784edc-26b1-480e-9349-95527305f12c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display, Markdown, HTML\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib_inline.backend_inline\n",
    "matplotlib_inline.backend_inline.set_matplotlib_formats(\"jpeg\")\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as pyplot\n",
    "import seaborn\n",
    "\n",
    "import numpy\n",
    "import polars\n",
    "import pandas\n",
    "import scipy\n",
    "\n",
    "import json\n",
    "\n",
    "polars.toggle_string_cache(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3de905-d38a-44a4-a969-006fb30b5ae3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Style Matplotlib plots\n",
    "\"\"\"\n",
    "\n",
    "plot_style = {\n",
    "    \"font.size\": 12,\n",
    "    \"font.family\": \"sans-serif\",\n",
    "    \"font.sans-serif\": [\"Inter\"],\n",
    "    \"figure.figsize\": [2, 2],\n",
    "    \"figure.dpi\": 150,\n",
    "    \"savefig.dpi\": 300,\n",
    "    \"figure.facecolor\": (1, 1, 1, 0),\n",
    "    \"text.usetex\": False,\n",
    "    \"lines.markersize\": 3,\n",
    "    \"axes.titleweight\": \"bold\",\n",
    "    \"axes.labelweight\": 600,\n",
    "    \"axes.labelsize\": 9,\n",
    "    \"axes.facecolor\": \"none\",\n",
    "    \"axes.spines.top\": False,\n",
    "    \"axes.spines.right\": False,\n",
    "    \"axes.linewidth\": 1.5,\n",
    "    \"grid.color\": \"#303030\",\n",
    "    \"grid.alpha\": 0.1,\n",
    "    \"xtick.labelsize\": 9,\n",
    "    \"xtick.major.width\": 1.5,\n",
    "    \"xtick.minor.width\": 1,\n",
    "    \"xtick.minor.visible\": False,\n",
    "    \"ytick.labelsize\": 9,\n",
    "    \"ytick.major.width\": 1.5,\n",
    "    \"ytick.minor.width\": 1,\n",
    "    \"ytick.minor.visible\": False,\n",
    "    \"figure.frameon\": False,\n",
    "    \"legend.frameon\": False,\n",
    "    \"legend.fancybox\": False,\n",
    "    \"legend.fontsize\": 9,\n",
    "    \"legend.scatterpoints\": 1,\n",
    "    \"legend.markerscale\": 1,\n",
    "    \"legend.handlelength\": 1.0,\n",
    "    \"legend.handletextpad\": 0.2,\n",
    "    \"axes.prop_cycle\": matplotlib.cycler(color=[\"k\", \"b\", \"g\", \"r\", \"c\", \"y\"]),\n",
    "}\n",
    "\n",
    "pyplot.style.use(plot_style)\n",
    "_ = pyplot.ioff()\n",
    "\n",
    "cm_white_to_gray = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_gray\", [\"#ffffff\", \"#404040\"])\n",
    "cm_white_to_red = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_red\", [\"#ffffff\", \"#ff0000\"])\n",
    "cm_white_to_blue = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_blue\", [\"#ffffff\", \"#0000ff\"])\n",
    "cm_white_to_green = matplotlib.colors.LinearSegmentedColormap.from_list(\"white_to_green\", [\"#ffffff\", \"#009000\"])\n",
    "\n",
    "cm_light_to_gray = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_gray\", [\"#e0e0e0\", \"#404040\"])\n",
    "cm_light_to_red = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_red\", [\"#ffe0e0\", \"#ff0000\"])\n",
    "cm_light_to_blue = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_blue\", [\"#e0e0ff\", \"#0000ff\"])\n",
    "cm_light_to_green = matplotlib.colors.LinearSegmentedColormap.from_list(\"light_to_green\", [\"#e0ffe0\", \"#009000\"])\n",
    "\n",
    "boxplot_style_args = {\n",
    "    \"whiskerprops\": {\n",
    "        \"color\": \"#000000\"\n",
    "    },\n",
    "    \"boxprops\": {\n",
    "        \"edgecolor\": \"#000000\",\n",
    "    },\n",
    "    \"flierprops\": {\n",
    "        \"marker\": \"o\",\n",
    "        \"markerfacecolor\": \"#000000\",\n",
    "        \"markersize\": 1,\n",
    "        \"markeredgecolor\": None\n",
    "    },\n",
    "    \"medianprops\": {\n",
    "        \"color\": \"#000000\"\n",
    "    },\n",
    "    \"capprops\": {\n",
    "        \"color\": \"#000000\"\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ce34945-48b6-431e-94b4-6819879fcb4c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Common plotting functions\n",
    "\"\"\"\n",
    "\n",
    "# Adapted from https://stackoverflow.com/a/53865762\n",
    "\n",
    "def calc_weights_for_scatter(x, y, bins=20):\n",
    "    if type(x) is not numpy.ndarray or type(y) is not numpy.ndarray:\n",
    "        raise ValueError(\"x and y must be `numpy.ndarray`s\")\n",
    "    \n",
    "    bin_weights, xbins, ybins = numpy.histogram2d(x, y, bins=bins, density=False)\n",
    "    weights = scipy.interpolate.interpn(points=(0.5*(xbins[1:] + xbins[:-1]), 0.5*(ybins[1:] + ybins[:-1])), values=bin_weights, xi=numpy.vstack([x, y]).T, method=\"splinef2d\", bounds_error=False)\n",
    "    weights[numpy.where(numpy.isnan(weights))] = 0.0\n",
    "    \n",
    "    normalizer = matplotlib.colors.Normalize(vmin=numpy.min(weights), vmax=numpy.max(weights))\n",
    "    \n",
    "    return weights, normalizer\n",
    "\n",
    "def reorder_points_and_weights(x, y, w, f = None):\n",
    "    if type(x) is not numpy.ndarray or type(y) is not numpy.ndarray or type(w) is not numpy.ndarray:\n",
    "        raise ValueError(\"x, y, and w must be `numpy.ndarray`s\")\n",
    "    \n",
    "    x_, y_, w_ = x, y, w\n",
    "    if f is not None:\n",
    "        x_, y_, w_ = f(x), f(y), f(w)\n",
    "    order = numpy.argsort(w_)\n",
    "    return {\n",
    "        \"x\": x_[order],\n",
    "        \"y\": y_[order],\n",
    "        \"c\": w_[order]\n",
    "    }\n",
    "\n",
    "NAME_TRANSLATION_TABLE = {\n",
    "    \"WT\": \"+/+\",\n",
    "    \"Rhet\": \"R90W/+\",\n",
    "    \"Rhom\": \"R90W/R90W\",\n",
    "    \"Ehet\": \"E168d2/+\",\n",
    "    \"Ehom\": \"E168d2/E168d2\",\n",
    "    \"CrxKO\": \"-/-\",\n",
    "    \"chip_peak_id\": \"CRE\"\n",
    "}\n",
    "\n",
    "def translate_genotype_to_label(genotype):\n",
    "    return NAME_TRANSLATION_TABLE[genotype]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36060300-fd46-4181-bf3d-7978c28f574d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Read in sample information and total read counts, filter to rows of interest\n",
    "\"\"\"\n",
    "\n",
    "samples_information = polars.scan_csv(\"sample_information.tsv\", sep=\"\\t\").with_columns([\n",
    "    polars.format(\"{}.{}.{}.batch{}.rep{}\", \"genotype\", \"library\", \"source_molecule\", \"sequencing_group\", \"replicate\").alias(\"computed_name\"),\n",
    "    polars.col(\"genotype\").cast(polars.Categorical),\n",
    "    polars.col(\"source_molecule\").cast(polars.Categorical),\n",
    "    polars.col(\"library\").cast(polars.Categorical),\n",
    "]).collect()\n",
    "\n",
    "display(Markdown(\"### Read File Information\"))\n",
    "display(samples_information)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6000faa0-6b28-4b8c-84ea-cebe75c3d928",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Load barcode counts for each sample\n",
    "\"\"\"\n",
    "\n",
    "raw_samples_counts = []\n",
    "\n",
    "for sample_info in tqdm(samples_information.iter_rows(named=True), total=52):\n",
    "    sample_counts = polars.read_csv(f\"\"\"Barcodes/{sample_info['computed_name']}.barcode_counts.tsv\"\"\", sep=\"\\t\")\n",
    "    \n",
    "    raw_samples_counts.append(sample_counts.with_columns([\n",
    "        polars.lit(polars.DataFrame(sample_info).to_struct('sample_info')).alias(\"sample_info\")\n",
    "    ]).unnest(\"sample_info\").with_columns([\n",
    "        polars.col(\"genotype\").cast(polars.Categorical),\n",
    "        polars.col(\"source_molecule\").cast(polars.Categorical),\n",
    "        polars.col(\"library\").cast(polars.Categorical),\n",
    "    ]))\n",
    "    \n",
    "samples_counts = polars.concat(raw_samples_counts, rechunk=True, how=\"vertical\", parallel=True)\n",
    "\n",
    "display(Markdown(\"### Sample Counts\"))\n",
    "display(samples_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f5d763c-fc18-4566-9629-b5a739127633",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "display(Markdown(f\"### Total Counts per Sample\"))\n",
    "grid = seaborn.catplot(kind=\"bar\", data=samples_counts.groupby([\"genotype\", \"source_molecule\", \"library\", \"replicate\"]).agg([polars.col(\"count\").sum()]).sort([\"library\", \"source_molecule\", \"genotype\"]).to_pandas(), row=\"library\", col=\"genotype\", x=\"replicate\", y=\"count\", order=[1, 2, 3], height=3, palette=\"Set1\")\n",
    "display(grid.figure) \n",
    "pyplot.close(grid.figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b75e8f9-c0a2-4eb2-953e-6e2fb4b5a3f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Load BC-library_id map\n",
    "\"\"\"\n",
    "\n",
    "bc_to_library_id_map = polars.read_csv(\"bc_to_library_id_map.tsv\", sep=\"\\t\")\n",
    "\n",
    "display(Markdown(\"### BC-to-Library ID Map\"))\n",
    "display(bc_to_library_id_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54dd99d7-9818-4efb-9b5b-a866a4f4f7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline = samples_counts.lazy()\n",
    "\n",
    "\"\"\"\n",
    "Sum barcode counts for samples sequenced in multiple runs (multiple sequencing_group values)\n",
    "\"\"\"\n",
    "\n",
    "pipeline = pipeline.groupby([\"library\", \"genotype\", \"source_molecule\", \"replicate\", \"BC\"]).agg([\n",
    "    polars.sum(\"count\")\n",
    "])\n",
    "\n",
    "\"\"\"\n",
    "Extract list of library members with zero DNA counts (not observed in sequencing of plasmid pool)\n",
    "\"\"\"\n",
    "\n",
    "missing_from_plasmid = pipeline.filter((polars.col(\"genotype\") == \"Plasmid\") & (polars.col(\"count\") == 0)).collect().groupby([\"library\"]).all().select([\"library\", \"BC\"]).with_columns([\n",
    "    polars.col(\"BC\").arr.lengths().alias(\"BC_count\")\n",
    "])\n",
    "display(Markdown(\"### BCs missing from Plasmid\"))\n",
    "display(missing_from_plasmid)\n",
    "\n",
    "\"\"\"\n",
    "Compute counts per million reads\n",
    "\"\"\"\n",
    "\n",
    "pipeline = pipeline.with_columns([\n",
    "    ((polars.col(\"count\") * 1_000_000) / polars.sum(\"count\")).alias(\"cpm\"),\n",
    "])\n",
    "\n",
    "\"\"\"\n",
    "Normalize cpm for each measurement by dividing by cpm in the plasmid (DNA) sample.\n",
    "Measurements for which plasmid (DNA) counts are 0 are dropped.\n",
    "\"\"\"\n",
    "\n",
    "pipeline = pipeline.join(\n",
    "    pipeline\n",
    "        .filter(polars.col(\"genotype\") == \"Plasmid\")\n",
    "        .select([\"library\", \"BC\", \"cpm\", \"count\"])\n",
    "        .rename({\"cpm\": \"plasmid_cpm\", \"count\": \"plasmid_count\"})\n",
    ", on=[\"library\", \"BC\"], how=\"left\").with_columns([\n",
    "    (polars.col(\"cpm\") / polars.col(\"plasmid_cpm\")).alias(\"activity\")\n",
    "]).filter((polars.col(\"genotype\") != \"Plasmid\") & (polars.col(\"plasmid_count\") > 0)).drop([\"source_molecule\", \"plasmid_cpm\", \"plasmid_count\"])\n",
    "\n",
    "\"\"\"\n",
    "Join counts with library ID map\n",
    "\"\"\"\n",
    "\n",
    "pipeline = pipeline.join(bc_to_library_id_map.lazy(), on=\"BC\", how=\"left\")\n",
    "\n",
    "pipeline = pipeline.sort([\"library\", \"genotype\", \"library_id\", \"BC\", \"replicate\"])\n",
    "\n",
    "normalized_samples_counts = pipeline.collect()\n",
    "\n",
    "display(Markdown(\"### Plasmid-normalized Counts\"))\n",
    "display(normalized_samples_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abe1a227-c220-4c5c-9d24-b983467802dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown(\"## basal (all barcodes and replicates)\"))\n",
    "data = normalized_samples_counts.filter(polars.col(\"library_id\") == \"basal\").to_pandas()\n",
    "data[\"genotype\"].cat = data[\"genotype\"].cat.remove_unused_categories()\n",
    "\n",
    "grid = seaborn.catplot(kind=\"strip\", data=data, row=\"library\", x=\"genotype\", hue=\"replicate\", y=\"count\", height=3, aspect=2, palette=\"Set1\", alpha=0.4, linewidth=0, dodge=True)\n",
    "\n",
    "for axis in grid.axes.flat:\n",
    "    axis.axhline(10, color=\"#00000080\", linestyle=\":\")\n",
    "    axis.set_yscale(\"symlog\", linthresh=1)\n",
    "    axis.set_ylim(bottom=0)\n",
    "\n",
    "display(grid.figure)\n",
    "pyplot.close(grid.figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02998a18-edbf-4286-8ee5-421ca3e6fc9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown(\"## scrambled sequences (all barcodes and replicates)\"))\n",
    "\n",
    "data = normalized_samples_counts.filter(polars.col(\"library_id\").str.contains(\"Scramble\")).to_pandas()\n",
    "data[\"genotype\"].cat = data[\"genotype\"].cat.remove_unused_categories()\n",
    "\n",
    "grid = seaborn.catplot(kind=\"strip\", data=data, row=\"library\", x=\"genotype\", y=\"count\", hue=\"replicate\", dodge=True, height=3, aspect=2, palette=\"Set1\", alpha=0.1, linewidth=0)\n",
    "\n",
    "for axis in grid.axes.flat:\n",
    "    axis.axhline(10, color=\"#00000080\", linestyle=\":\")\n",
    "    axis.set_yscale(\"symlog\", linthresh=10)\n",
    "    axis.set_ylim(bottom=0)\n",
    "\n",
    "display(grid.figure)\n",
    "pyplot.close(grid.figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37a38234-38cd-4103-bc39-b2ed413647b8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_r2s = {}\n",
    "\n",
    "for (library, genotype), dataframe in normalized_samples_counts.sort([\"library\", \"genotype\"]).groupby([\"library\", \"genotype\"], maintain_order=True):\n",
    "    display(Markdown(f\"## {genotype} ({library})\"))\n",
    "    data = dataframe.with_columns([\n",
    "        polars.col(\"activity\") + 1e-4\n",
    "    ]).pivot(values=\"activity\", index=\"library_id\", columns=\"replicate\", aggregate_function=\"mean\")\n",
    "    \n",
    "    figure, axes = pyplot.subplots(nrows=1, ncols=3, sharex=False, sharey=False, figsize=(5.25, 1.75))\n",
    "    \n",
    "    comparisons = [(\"1\", \"2\"), (\"1\", \"3\"), (\"2\", \"3\")]\n",
    "    \n",
    "    bounds = (1e-5, 1e2)\n",
    "    \n",
    "    for index, ((rep1, rep2), axis) in enumerate(zip(comparisons, axes.flat)):\n",
    "        x = data[rep1].to_numpy()\n",
    "        y = data[rep2].to_numpy()\n",
    "        \n",
    "        bins = numpy.geomspace(start=bounds[0], stop=bounds[1], num=150)\n",
    "        \n",
    "        # Calculate R^2\n",
    "        fit = scipy.stats.linregress(numpy.log10(x), numpy.log10(y))\n",
    "        fit_func = lambda x: fit.slope * x + fit.intercept\n",
    "\n",
    "        if library not in all_r2s:\n",
    "            all_r2s[library] = []\n",
    "        all_r2s[library].append(fit.rvalue**2)\n",
    "        \n",
    "        # Plot wildtype CREs as a scatterplot colored by density\n",
    "        weights, color_normalizer = calc_weights_for_scatter(x, y, bins=bins)\n",
    "        axis.scatter(**reorder_points_and_weights(x, y, weights), cmap=cm_light_to_gray, norm=color_normalizer, s=6, marker=\"o\", rasterized=True, zorder=1, label=f\"RÂ²={fit.rvalue**2:.2f}\")\n",
    "        \n",
    "        axis.set(xlim=bounds, ylim=bounds, xlabel=f\"Rep{rep1}\", ylabel=f\"Rep{rep2}\")\n",
    "        axis.set_xscale(\"log\", base=10)\n",
    "        axis.set_yscale(\"log\", base=10)\n",
    "        axis.tick_params(which=\"major\", bottom=True, top=False, left=True, right=False, labelbottom=False, labeltop=False, labelleft=False, labelright=False)\n",
    "        axis.tick_params(which=\"minor\", bottom=False, top=False, left=False, right=False)\n",
    "\n",
    "        axis.set_aspect(aspect=1, adjustable=\"box\", anchor=\"C\")\n",
    "        \n",
    "        handles, labels = axis.get_legend_handles_labels()\n",
    "        legend = axis.legend(handles, labels, loc=\"upper left\")\n",
    "\n",
    "    for axis in axes.flat[len(comparisons):]:\n",
    "        axis.axis(\"off\")\n",
    "\n",
    "    figure.tight_layout()\n",
    "    figure.savefig(f\"Reproduciblilty/reproducibility_{library}_{genotype}.pdf\", bbox_inches=\"tight\")\n",
    "    display(figure)\n",
    "    pyplot.close(figure)\n",
    "    \n",
    "for library, r2_values in all_r2s.items():\n",
    "    display(Markdown(f\"#### {library} avg. R^2\"))\n",
    "    print(\"{:.2f}\".format(numpy.mean(r2_values)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0d3ea1c-3fe5-4e44-b29c-1e7053334b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline = normalized_samples_counts.lazy()\n",
    "    \n",
    "\"\"\"\n",
    "Aggregate counts for each library member (CRE) across all barcodes and replicates; compute summary statistics\n",
    "\"\"\"\n",
    "\n",
    "pipeline = pipeline.groupby([\"library\", \"genotype\", \"library_id\"]).agg([\n",
    "    polars.col(\"activity\").mean().alias(\"activity_mean\"),\n",
    "    polars.col(\"activity\").std().alias(\"activity_std\"),\n",
    "    polars.count().alias(\"n_observations\"),\n",
    "]).with_columns([\n",
    "    (polars.col(\"activity_std\") / polars.col(\"activity_mean\")).alias(\"activity_cov\")\n",
    "]).with_columns([\n",
    "    (polars.col(\"activity_mean\") / (polars.col(\"activity_cov\").pow(2) + 1).sqrt()).log().alias(\"activity_mu\"),\n",
    "    (polars.col(\"activity_cov\").pow(2) + 1).log().sqrt().alias(\"activity_sigma\"),\n",
    "])\n",
    "\n",
    "\"\"\"\n",
    "Within each genotype, calculate t-tests for each CRE vs basal\n",
    "\"\"\"\n",
    "\n",
    "def _calculate_t_tests(arguments):\n",
    "    results = []\n",
    "    for i in range(len(arguments[0])):\n",
    "        results.append(scipy.stats.ttest_ind_from_stats(*map(lambda l: l[i], arguments), equal_var=False)[1])\n",
    "    return polars.Series(\"pvalue\", results)\n",
    "\n",
    "pipeline = pipeline.join(\n",
    "    pipeline.filter(polars.col(\"library_id\") == \"basal\").select([\n",
    "        polars.col(\"library\"),\n",
    "        polars.col(\"genotype\"),\n",
    "        polars.col(\"activity_mean\").prefix(\"basal_\"),\n",
    "        polars.col(\"activity_mu\").prefix(\"basal_\"),\n",
    "        polars.col(\"activity_sigma\").prefix(\"basal_\"),\n",
    "        polars.col(\"n_observations\").prefix(\"basal_\")\n",
    "    ]),\n",
    "    on=[\"library\", \"genotype\"], how=\"left\"\n",
    ").with_columns([\n",
    "    polars.map([\"basal_activity_mu\", \"basal_activity_sigma\", \"basal_n_observations\", \"activity_mu\", \"activity_sigma\", \"n_observations\"], _calculate_t_tests).fill_nan(1).alias(\"pvalue\")\n",
    "]).with_columns([\n",
    "    (polars.col(\"pvalue\") * (polars.col(\"pvalue\").count() / polars.col(\"pvalue\").rank(method=\"max\"))).alias(\"qvalue\")\n",
    "])\n",
    "\n",
    "\"\"\"\n",
    "Compute 5th and 95 %ile of scrambled library members within each genotype; classify CREs\n",
    "\"\"\"\n",
    "\n",
    "pipeline = pipeline.join(\n",
    "    pipeline.filter(polars.col(\"library_id\").str.contains(\"Scramble\")).groupby([\"library\", \"genotype\"]).agg([\n",
    "        polars.col(\"activity_mean\").apply(lambda series: numpy.percentile(series, 5)).alias(\"scrambled_5th_percentile\"),\n",
    "        polars.col(\"activity_mean\").apply(lambda series: numpy.percentile(series, 20)).alias(\"scrambled_20th_percentile\"),\n",
    "        polars.col(\"activity_mean\").apply(lambda series: numpy.percentile(series, 80)).alias(\"scrambled_80th_percentile\"),\n",
    "        polars.col(\"activity_mean\").apply(lambda series: numpy.percentile(series, 95)).alias(\"scrambled_95th_percentile\")\n",
    "    ]),\n",
    "    on=[\"library\", \"genotype\"], how=\"left\"\n",
    ").with_columns([\n",
    "    polars.when((polars.col(\"activity_mean\") > polars.col(\"scrambled_95th_percentile\")) & (polars.col(\"qvalue\") < 0.05)).then(\"strong_enhancer\")\n",
    "        .when((polars.col(\"activity_mean\") > polars.col(\"basal_activity_mean\")) & (polars.col(\"qvalue\") < 0.05)).then(\"weak_enhancer\")\n",
    "        .when((polars.col(\"activity_mean\") < polars.col(\"scrambled_5th_percentile\"))& (polars.col(\"qvalue\") < 0.05)).then(\"strong_silencer\")\n",
    "        .when((polars.col(\"activity_mean\") < polars.col(\"basal_activity_mean\")) & (polars.col(\"qvalue\") < 0.05)).then(\"weak_silencer\")\n",
    "        .otherwise(\"inactive\").cast(polars.Categorical).alias(\"activity_class\")\n",
    "]).drop([\n",
    "    \"scrambled_5th_percentile\",\n",
    "    \"scrambled_20th_percentile\",\n",
    "    \"scrambled_80th_percentile\",\n",
    "    \"scrambled_95th_percentile\",\n",
    "    \"activity_cov\",\n",
    "    \"basal_activity_mean\",\n",
    "    \"basal_activity_mu\",\n",
    "    \"basal_activity_sigma\",\n",
    "    \"basal_n_observations\"\n",
    "])\n",
    "\n",
    "pipeline = pipeline.sort([\"library\", \"library_id\", \"genotype\"])\n",
    "\n",
    "processed_samples_counts = pipeline.collect()\n",
    "\n",
    "display(Markdown(\"### Aggregated Normalized Counts\"))\n",
    "display(processed_samples_counts)\n",
    "\n",
    "\"\"\"\n",
    "Write out processed counts\n",
    "\"\"\"\n",
    "\n",
    "processed_samples_counts.write_parquet(\"processed_library_counts.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "affa3a45-0829-4f95-99ec-7cdb7ae05cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown(\"## scrambled sequences (means across replicates and barcodes)\"))\n",
    "\n",
    "data = processed_samples_counts.filter(polars.col(\"library_id\").str.contains(\"Scramble\")).to_pandas()\n",
    "data[\"genotype\"].cat = data[\"genotype\"].cat.remove_unused_categories()\n",
    "\n",
    "grid = seaborn.catplot(kind=\"strip\", data=data, row=\"library\", x=\"genotype\", hue=\"genotype\", y=\"activity_mean\", height=3, aspect=2, palette=\"Set1\", alpha=0.4, linewidth=0)\n",
    "for axis in grid.axes.flat:\n",
    "    axis.set_yscale(\"symlog\", linthresh=1e-1)\n",
    "    axis.set_ylim(bottom=0)\n",
    "display(grid.figure)\n",
    "pyplot.close(grid.figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339be2ed-7231-417c-820a-598a9b36d8e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
